{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "XQOfEBWaHOnK"
      },
      "source": [
        "**Author:** Pratik Vyas\n",
        "\n",
        "**Classification:** Binary classification ( Cancer , non-Cancer)\n",
        "\n",
        "**Usecase:** Distributed Finetuning LLM 'meta-llama/Llama-3.1-8B-Instruct' with LORA"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "collapsed": true,
        "id": "oli34wRCkbWR",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "fb5b4385-69cc-4879-9889-741cdf67fec8"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m10.5/10.5 MB\u001b[0m \u001b[31m35.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m363.4/363.4 MB\u001b[0m \u001b[31m2.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m13.8/13.8 MB\u001b[0m \u001b[31m40.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m24.6/24.6 MB\u001b[0m \u001b[31m28.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m883.7/883.7 kB\u001b[0m \u001b[31m38.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m664.8/664.8 MB\u001b[0m \u001b[31m2.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m211.5/211.5 MB\u001b[0m \u001b[31m5.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m56.3/56.3 MB\u001b[0m \u001b[31m10.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m127.9/127.9 MB\u001b[0m \u001b[31m7.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m207.5/207.5 MB\u001b[0m \u001b[31m5.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m21.1/21.1 MB\u001b[0m \u001b[31m52.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m67.0/67.0 MB\u001b[0m \u001b[31m9.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m310.2/310.2 kB\u001b[0m \u001b[31m10.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m491.5/491.5 kB\u001b[0m \u001b[31m29.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m193.6/193.6 kB\u001b[0m \u001b[31m19.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
            "gcsfs 2025.3.2 requires fsspec==2025.3.2, but you have fsspec 2025.3.0 which is incompatible.\u001b[0m\u001b[31m\n",
            "\u001b[0m"
          ]
        }
      ],
      "source": [
        "!pip3 install -q -U transformers accelerate\n",
        "!pip3 install -q -U bitsandbytes\n",
        "!pip3 install -q -U wandb\n",
        "# !pip3 install -q  -U trl\n",
        "!pip3 install -q  \"trl == 0.12.0\""
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {
        "id": "BO1RnxRTlfLf"
      },
      "outputs": [],
      "source": [
        "from transformers import AutoTokenizer, AutoModelForCausalLM, BitsAndBytesConfig, pipeline"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {
        "id": "FWcJLMK3kvN4"
      },
      "outputs": [],
      "source": [
        "import os\n",
        "\n",
        "os.environ[\"HF_TOKEN\"] =\"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\"\n",
        "os.environ[\"WB_KEY\"] = \"9d7decf681236b200a35c0121bca0fe725be724c\""
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 272,
          "referenced_widgets": [
            "1123a51845ce43ba8164dbc1a0c4a050",
            "0ecc0c7a32634a9e8f4a121fb0a8e88b",
            "8d3668193295457f8c3869cf11e5374a",
            "564fc202d28c4a0187b9ddc1714f5dff",
            "2b9355e53e4a4a308e7dfcc4766f4e68",
            "b1df748eb36f495fbecc7606fe0ab4ab",
            "2afe361e314c40fc82c06f9c05b0bd78",
            "503e6d9e50a24990a70b6ac36d126e4d",
            "6f7fd6447df04125a96c956a8f584a91",
            "1d996e9851544692a19b24a98fd1051b",
            "dc36607256984b06832366346058990e",
            "c4188699ad52466dbd0fdbbbf100dfd0",
            "909f67ced1414c30958dfac8a46789e7",
            "6bda810492ce4d39a8baa4ee23009394",
            "4c25bf62179d4fb8b845e61a13ad42a3",
            "ff40e3c4f54b49ef9ee1b5f30157068e",
            "935f039e6027407d9bf8fc3c83542023",
            "964bf06aa20047e38fafd593e5fb5c51",
            "0f3a314f3677424f9942a195f1c17f36",
            "bb565a584f78435ebd64966e19921bbd",
            "38fe86dfd44f44888aabadcadde5ca50",
            "1389d3d4be354fac8d2535e25717e3d7",
            "34894031b103496e908a9b777831eff7",
            "2b0a9899f60b4108a6c3efc668065a03",
            "0868440e6e21457e86777b058b2d036a",
            "cf938b70e50541ff8af19636de10fe72",
            "708f97ac003849158105810f3b4070e3",
            "f1114a36bf2840ccaadb1c6f91850feb",
            "16637baf400a45b0bd631300e5148bb4",
            "125e3734b1784c3c97de9d9e29b1f884",
            "8794c6dbb78747d5856efc367fbfb937",
            "41815a9d6b804df88c0e4e8a2078a271",
            "49bffe2771114187bb49d47a06468e4c"
          ]
        },
        "id": "EMUpbmcXlFw_",
        "outputId": "ead009b3-ce8b-4a65-de67-f3caff5e3093"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.11/dist-packages/transformers/models/auto/tokenization_auto.py:902: FutureWarning: The `use_auth_token` argument is deprecated and will be removed in v5 of Transformers. Please use `token` instead.\n",
            "  warnings.warn(\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "tokenizer_config.json:   0%|          | 0.00/55.4k [00:00<?, ?B/s]"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "1123a51845ce43ba8164dbc1a0c4a050"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: \n",
            "The secret `HF_TOKEN` does not exist in your Colab secrets.\n",
            "To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.\n",
            "You will be able to reuse this secret in all of your notebooks.\n",
            "Please note that authentication is recommended but still optional to access public models or datasets.\n",
            "  warnings.warn(\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "tokenizer.json:   0%|          | 0.00/9.09M [00:00<?, ?B/s]"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "c4188699ad52466dbd0fdbbbf100dfd0"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "special_tokens_map.json:   0%|          | 0.00/296 [00:00<?, ?B/s]"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "34894031b103496e908a9b777831eff7"
            }
          },
          "metadata": {}
        }
      ],
      "source": [
        "model_id=\"meta-llama/Llama-3.1-8B-Instruct\"\n",
        "new_model=\"Llama-3.1-8B-Inst_cancer_classification_finetuned\"\n",
        "\n",
        "\n",
        "tokenizer = AutoTokenizer.from_pretrained(model_id,   use_auth_token=os.environ[\"HF_TOKEN\"])"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 499,
          "referenced_widgets": [
            "b921eda48a784a16a1ab23c1114ee3b8",
            "5ec0d973e8814e8cba7e5800897fca8e",
            "9c197eb30d2b40e69644ba946eff98ea",
            "2b338f0c319046e1bdd9aaabfa6eaa99",
            "34c85947a32e441385f27c12c8db595d",
            "0fc1cf1c39ed47959e90e961ccdb31c4",
            "0068ae4266484b43bab4025a40cb6c2a",
            "13b89dc0cf8b47b29198d02f6f709d2e",
            "79b0046fcd9f46a3aaa5bce0ed5ef6d0",
            "24bda1092d7a470d9ea9c6bb3dd2f8ff",
            "ad716a9966854e8380e04c509bc076d9",
            "772566c131e345ddb3c98da8493b270c",
            "a2f23560d4624c43a459a8540cabec3c",
            "6bc15a049e8942eda8a2ca39342f4e39",
            "f8b6e70d4be5457da5a9fe20742fee81",
            "cf4f51fc74574c1ead8dbce01048cb27",
            "d168a5f4366c4c119cee76111c513c16",
            "f568d4c2e891417aa9d51ebe8f760fe1",
            "c563e276d6994bfd8ec7ee63c65c6f43",
            "ac4891ae99ac4a58974aec4c0d1a96a8",
            "5a31d0aac7294f1cbeebb2a24e0f3085",
            "49bbf614a8754f42be8c4899166ce3a4",
            "6d02dd36153a4567be6a772bc8524ace",
            "fb73a2968dbf4ebe87ac3daa01ad4f08",
            "a55d5960bd9a42c081824fed93d94407",
            "c6f77dd4b9a846ceb9f12255b8094b44",
            "cf6dbe9d98ec494cba8d679da1d6798b",
            "1a5161a5338940debe57ca9e49ee2367",
            "a136eaa12dcd4d999097a2d260245fd0",
            "3bc919535d704f34816ba838b11234d9",
            "0d6522d891db4386bc38fbac5b4dc5c2",
            "7462363d1c4f426998b6a99e6709dd2e",
            "1786368e14074e4c9fa187574a20e3fe",
            "744f034269e34f61bb70581da1789452",
            "98e4c74ee7cb4de69a8a232aec0e202f",
            "ad0efb35a66245ab992dee8a6733e1d0",
            "85d8dcbfef51406496d0895308f610fb",
            "fffe2b2fb69444319ed6833a12e3a79f",
            "5f326d70259b4f41af9f10dc5954770b",
            "7304c549bc6447deb2d75aff7b4d72c3",
            "25e49a99eb2346839db6243bf8ce9cec",
            "9bca9b384e35486194cf649bf9ab6953",
            "c6779309b2344c8ba3438f6a5910ffbc",
            "5fe18beae0a644bcbea8e5a0de3ac208",
            "3e24efdc56864816953782193d905806",
            "2e7123a6cd2546f8b9ee8f97f9877429",
            "a382450c4ec24d76acc8bce002aa86b2",
            "713d25dc645f462782016f498946469d",
            "490fe97e5eec4f33b26c61b51b76c3fa",
            "b640de23a2944bd180c6f8658cbdc9be",
            "30b1ff4804484c3da5e658a8c9c95fc9",
            "50c3f8f1a9644c55a046a8b38c1d4dc5",
            "3abcd2db1a4e4d7babff16a619fc749f",
            "c4af8a3a3fcc4f2db81b401e18e23243",
            "82e471790f7f465aac79565993626eaf",
            "615305d215d74b90a6dfe54c82771875",
            "9e8a415be34942b5b5e6dcedda227009",
            "35d6c77b4ae0404496e944fe285fa1dc",
            "df92350bd8974429b814112ea846fe15",
            "0e2911af258640ff8c4e117261ef6145",
            "e90f2717403c4e7ea2d8be5110d91898",
            "3463ee3e462c4f2b93844d234375a8b3",
            "df09b0309a1c4c55bb31a1ea6148482c",
            "b47e3d2361f041ad9abdf34ea7971410",
            "fce17887532b4cdca0d004b1f2c64b54",
            "7d6a3e4a0c014985b82044322bcadca9",
            "a492cae1ecc94d95a4e92cb33243985d",
            "139c697ef3f348a5b8647d682aaa5130",
            "0131779bee3d457f96d3c1c52c3336d5",
            "906a6cd3528f451a8af16cb85ca3f253",
            "0b684e88b5904de4bf43564442fda998",
            "b438afc27694436ba5fd762c77cafe97",
            "7e6c8a2586e343e5b255c9882ed19a69",
            "33f260ee397348589841ca2a92d7ad06",
            "ad59653e22ef44f0b242bf12a19d9054",
            "f6a4f41cb61945c38816d504cdeb8aa2",
            "f7c9919821aa4559b33bd92a18085b57",
            "a4d62e58e2ed4730b240b51967d6367f",
            "a3a79f714fcc430185097786633794cd",
            "97590e7d765d430497923e36c75ee44e",
            "a4d47db45f23481c852fe5c804a81c9a",
            "e26ecd03c06d4482b5a77fdf6c354491",
            "3d5825962891430e9f25a7394bb10077",
            "98f8ef359d4844b0a8740ff02e8ecd97",
            "4109b8d89866490cabd7c5c30567b1f1",
            "0fdd5dc6792147a9a64eb4903a542969",
            "39a8be84e9bb445cb5d07becebaa28c5",
            "96b53899ece440c39496367b6fe5fce7",
            "7447e4493ad34038842fe1e551672aa9",
            "4c8f7d656dff4d0697b8816b4f8b8cb6",
            "fa7fba46077746f396d19a835454f9d9",
            "75eb1fb98e3447b9b796bec427e6303e",
            "e41e09b76eb14d779db99a86c7dbcfe2",
            "59ff9bc159b04720814b474b1efa8621",
            "e747dfdb256a469e87b8f73135583209",
            "b58fa77c832a457caf2cf27c0768d770",
            "2ea087fa546a40dcb54299130040863e",
            "39f5714f432241c994a2b94dd9d5578e",
            "ff52a18074994489b4380f7c5bad842b"
          ]
        },
        "id": "7PQxBQ3yl7wq",
        "outputId": "8db687ad-4c18-4c1e-e0f6-b351b3fc4d9d"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.11/dist-packages/transformers/models/auto/auto_factory.py:476: FutureWarning: The `use_auth_token` argument is deprecated and will be removed in v5 of Transformers. Please use `token` instead.\n",
            "  warnings.warn(\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "config.json:   0%|          | 0.00/855 [00:00<?, ?B/s]"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "b921eda48a784a16a1ab23c1114ee3b8"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "model.safetensors.index.json:   0%|          | 0.00/23.9k [00:00<?, ?B/s]"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "772566c131e345ddb3c98da8493b270c"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "Fetching 4 files:   0%|          | 0/4 [00:00<?, ?it/s]"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "6d02dd36153a4567be6a772bc8524ace"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`\n",
            "WARNING:huggingface_hub.file_download:Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`\n",
            "Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`\n",
            "WARNING:huggingface_hub.file_download:Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`\n",
            "Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`\n",
            "WARNING:huggingface_hub.file_download:Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`\n",
            "Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`\n",
            "WARNING:huggingface_hub.file_download:Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "model-00004-of-00004.safetensors:   0%|          | 0.00/1.17G [00:00<?, ?B/s]"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "744f034269e34f61bb70581da1789452"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "model-00001-of-00004.safetensors:   0%|          | 0.00/4.98G [00:00<?, ?B/s]"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "3e24efdc56864816953782193d905806"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "model-00002-of-00004.safetensors:   0%|          | 0.00/5.00G [00:00<?, ?B/s]"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "615305d215d74b90a6dfe54c82771875"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "model-00003-of-00004.safetensors:   0%|          | 0.00/4.92G [00:00<?, ?B/s]"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "a492cae1ecc94d95a4e92cb33243985d"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "Loading checkpoint shards:   0%|          | 0/4 [00:00<?, ?it/s]"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "a4d62e58e2ed4730b240b51967d6367f"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "generation_config.json:   0%|          | 0.00/184 [00:00<?, ?B/s]"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "7447e4493ad34038842fe1e551672aa9"
            }
          },
          "metadata": {}
        }
      ],
      "source": [
        "\n",
        "# Activate 4-bit precision base model loading\n",
        "bnb_config = BitsAndBytesConfig(\n",
        "    load_in_4bit=True,\n",
        "    bnb_4bit_quant_type=\"nf4\", # Quantization type (fp4 or nf4)\n",
        "    bnb_4bit_compute_dtype=\"float16\",  # Compute dtype for 4-bit base models\n",
        "    bnb_4bit_use_double_quant=False,\n",
        "    # Enable CPU offloading for specific layers\n",
        "    llm_int8_enable_fp32_cpu_offload=False, # Activate nested quantization for 4-bit base models (double quantization)\n",
        ")\n",
        "\n",
        "\n",
        "\n",
        "##load model\n",
        "model = AutoModelForCausalLM.from_pretrained(\n",
        "    model_id,\n",
        "    quantization_config=bnb_config,\n",
        "    device_map=\"auto\",  # Let Transformers automatically decide device placement\n",
        "    use_auth_token=os.environ[\"HF_TOKEN\"]\n",
        ")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "mNnxjP8IwgFW",
        "outputId": "2cb1c9df-7623-4fd3-f43e-fc3cd6240b4f"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "LlamaForCausalLM(\n",
            "  (model): LlamaModel(\n",
            "    (embed_tokens): Embedding(128256, 4096)\n",
            "    (layers): ModuleList(\n",
            "      (0-31): 32 x LlamaDecoderLayer(\n",
            "        (self_attn): LlamaAttention(\n",
            "          (q_proj): Linear4bit(in_features=4096, out_features=4096, bias=False)\n",
            "          (k_proj): Linear4bit(in_features=4096, out_features=1024, bias=False)\n",
            "          (v_proj): Linear4bit(in_features=4096, out_features=1024, bias=False)\n",
            "          (o_proj): Linear4bit(in_features=4096, out_features=4096, bias=False)\n",
            "        )\n",
            "        (mlp): LlamaMLP(\n",
            "          (gate_proj): Linear4bit(in_features=4096, out_features=14336, bias=False)\n",
            "          (up_proj): Linear4bit(in_features=4096, out_features=14336, bias=False)\n",
            "          (down_proj): Linear4bit(in_features=14336, out_features=4096, bias=False)\n",
            "          (act_fn): SiLU()\n",
            "        )\n",
            "        (input_layernorm): LlamaRMSNorm((4096,), eps=1e-05)\n",
            "        (post_attention_layernorm): LlamaRMSNorm((4096,), eps=1e-05)\n",
            "      )\n",
            "    )\n",
            "    (norm): LlamaRMSNorm((4096,), eps=1e-05)\n",
            "    (rotary_emb): LlamaRotaryEmbedding()\n",
            "  )\n",
            "  (lm_head): Linear(in_features=4096, out_features=128256, bias=False)\n",
            ")\n"
          ]
        }
      ],
      "source": [
        "print(model)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 208
        },
        "id": "R5JNOdUyqGtk",
        "outputId": "42ff313b-b099-423b-cd71-bf7f0701a94c"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "\u001b[34m\u001b[1mwandb\u001b[0m: \u001b[33mWARNING\u001b[0m If you're specifying your api key in code, ensure this code is not shared publicly.\n",
            "\u001b[34m\u001b[1mwandb\u001b[0m: \u001b[33mWARNING\u001b[0m Consider setting the WANDB_API_KEY environment variable, or running `wandb login` from the command line.\n",
            "\u001b[34m\u001b[1mwandb\u001b[0m: No netrc file found, creating one.\n",
            "\u001b[34m\u001b[1mwandb\u001b[0m: Appending key for api.wandb.ai to your netrc file: /root/.netrc\n",
            "\u001b[34m\u001b[1mwandb\u001b[0m: Currently logged in as: \u001b[33mpratik_ai\u001b[0m to \u001b[32mhttps://api.wandb.ai\u001b[0m. Use \u001b[1m`wandb login --relogin`\u001b[0m to force relogin\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "Tracking run with wandb version 0.19.11"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "Run data is saved locally in <code>/content/wandb/run-20250601_121041-ko349hw3</code>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "Syncing run <strong><a href='https://wandb.ai/pratik_ai/cancer_classification/runs/ko349hw3?apiKey=9d7decf681236b200a35c0121bca0fe725be724c' target=\"_blank\">pretty-star-33</a></strong> to <a href='https://wandb.ai/pratik_ai/cancer_classification?apiKey=9d7decf681236b200a35c0121bca0fe725be724c' target=\"_blank\">Weights & Biases</a> (<a href='https://wandb.me/developer-guide' target=\"_blank\">docs</a>)<br>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              " View project at <a href='https://wandb.ai/pratik_ai/cancer_classification?apiKey=9d7decf681236b200a35c0121bca0fe725be724c' target=\"_blank\">https://wandb.ai/pratik_ai/cancer_classification?apiKey=9d7decf681236b200a35c0121bca0fe725be724c</a>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              " View run at <a href='https://wandb.ai/pratik_ai/cancer_classification/runs/ko349hw3?apiKey=9d7decf681236b200a35c0121bca0fe725be724c' target=\"_blank\">https://wandb.ai/pratik_ai/cancer_classification/runs/ko349hw3?apiKey=9d7decf681236b200a35c0121bca0fe725be724c</a>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "Do NOT share these links with anyone. They can be used to claim your runs."
            ]
          },
          "metadata": {}
        }
      ],
      "source": [
        "# integrate Weights & Biases (W&B) with training process for tracking, monitoring, and collaboration\n",
        "import os\n",
        "import wandb\n",
        "\n",
        "wandb.login(key=os.environ[\"WB_KEY\"])\n",
        "run = wandb.init(\n",
        "    project=\"cancer_classification\",\n",
        "    job_type=\"training\",\n",
        "    anonymous=\"allow\",\n",
        ")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {
        "id": "AxQkXxoDqsO2"
      },
      "outputs": [],
      "source": [
        "\n",
        "from transformers import (AutoModelForCausalLM,\n",
        "                          AutoTokenizer,\n",
        "                          BitsAndBytesConfig,\n",
        "                          TrainingArguments,\n",
        "                          pipeline,\n",
        "                          logging)\n",
        "from sklearn.metrics import (accuracy_score,\n",
        "                             classification_report,\n",
        "                             confusion_matrix)\n",
        "from sklearn.model_selection import train_test_split\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "import os\n",
        "from tqdm import tqdm\n",
        "import bitsandbytes as bnb\n",
        "import torch\n",
        "import torch.nn as nn\n",
        "import transformers\n",
        "from datasets import Dataset\n",
        "from peft import LoraConfig, PeftConfig\n",
        "from trl import SFTTrainer\n",
        "from trl import setup_chat_format"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "WgzMmEEHrMXf"
      },
      "source": [
        "# Load, preprocess cancer data from zip file"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {
        "id": "9riHoDzGrJo4"
      },
      "outputs": [],
      "source": [
        "import zipfile\n",
        "import pandas as pd\n",
        "import io\n",
        "import os\n",
        "\n",
        "def load_text_files_from_zip_to_dataframe(zip_path, directory_in_zip, **pandas_kwargs):\n",
        "    \"\"\"\n",
        "    Loads text files from a specific directory inside a ZIP archive into a Pandas DataFrame.\n",
        "    \"\"\"\n",
        "    data = []\n",
        "    filenames = []\n",
        "    try:\n",
        "        with zipfile.ZipFile(zip_path, 'r') as zip_ref:\n",
        "            for filename in zip_ref.namelist():\n",
        "                if filename.startswith(directory_in_zip) and filename.endswith('.txt'):  # Only process text files\n",
        "                    try:\n",
        "                        with zip_ref.open(filename) as text_file:\n",
        "                            content = text_file.read().decode('utf-8', errors='ignore').strip() # Read file contents, decode from bytes\n",
        "                            data.append(content)\n",
        "                            # Extract the filename without the directory path and extension\n",
        "                            file_no_path = os.path.basename(filename)\n",
        "                            file_no_ext, _ = os.path.splitext(file_no_path)\n",
        "                            filenames.append(file_no_ext)\n",
        "                    except Exception as e:\n",
        "                        print(f\"Error reading file {filename}: {e}\")\n",
        "                        continue # Skip to the next file\n",
        "\n",
        "        if not data:\n",
        "            print(f\"No text files found in directory: {directory_in_zip}\")\n",
        "            return None\n",
        "\n",
        "        df = pd.DataFrame({'texts': data, 'filename': filenames})  # Create DataFrame\n",
        "        return df\n",
        "\n",
        "    except FileNotFoundError:\n",
        "        print(f\"Error: ZIP file not found: {zip_path}\")\n",
        "        return None\n",
        "    except zipfile.BadZipFile as e:\n",
        "        print(f\"Error: Invalid ZIP file: {e}\")\n",
        "        return None\n",
        "    except Exception as e:\n",
        "        print(f\"An error occurred: {e}\")\n",
        "        return None"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 395
        },
        "id": "0NLa2r9wrYcf",
        "outputId": "02a99f99-f51f-47ea-9189-29a33d42202c"
      },
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "                                               texts  filename  labels\n",
              "0  <ID:31055803>\\nTitle: [Analysis of age-specifi...  31055803    True\n",
              "1  <ID:31164412>\\nTitle: T-Cell Deletion of MyD88...  31164412    True\n",
              "2  <ID:31094905>\\nTitle: MYCN Amplified Relapse F...  31094905    True\n",
              "3  <ID:31498304>\\nTitle: In Vivo Inhibition of Mi...  31498304    True\n",
              "4  <ID:30897768>\\nTitle: Breast Cancer and miR-SN...  30897768    True"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-473d1319-32e0-4e92-9fe9-84760b4ebfe4\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>texts</th>\n",
              "      <th>filename</th>\n",
              "      <th>labels</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>&lt;ID:31055803&gt;\\nTitle: [Analysis of age-specifi...</td>\n",
              "      <td>31055803</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>&lt;ID:31164412&gt;\\nTitle: T-Cell Deletion of MyD88...</td>\n",
              "      <td>31164412</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>&lt;ID:31094905&gt;\\nTitle: MYCN Amplified Relapse F...</td>\n",
              "      <td>31094905</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>&lt;ID:31498304&gt;\\nTitle: In Vivo Inhibition of Mi...</td>\n",
              "      <td>31498304</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>&lt;ID:30897768&gt;\\nTitle: Breast Cancer and miR-SN...</td>\n",
              "      <td>30897768</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-473d1319-32e0-4e92-9fe9-84760b4ebfe4')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-473d1319-32e0-4e92-9fe9-84760b4ebfe4 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-473d1319-32e0-4e92-9fe9-84760b4ebfe4');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    <div id=\"df-99f00d7d-5c25-441e-aeb2-719783aa7354\">\n",
              "      <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-99f00d7d-5c25-441e-aeb2-719783aa7354')\"\n",
              "                title=\"Suggest charts\"\n",
              "                style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "      </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "      <script>\n",
              "        async function quickchart(key) {\n",
              "          const quickchartButtonEl =\n",
              "            document.querySelector('#' + key + ' button');\n",
              "          quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "          quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "          try {\n",
              "            const charts = await google.colab.kernel.invokeFunction(\n",
              "                'suggestCharts', [key], {});\n",
              "          } catch (error) {\n",
              "            console.error('Error during call to suggestCharts:', error);\n",
              "          }\n",
              "          quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "          quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "        }\n",
              "        (() => {\n",
              "          let quickchartButtonEl =\n",
              "            document.querySelector('#df-99f00d7d-5c25-441e-aeb2-719783aa7354 button');\n",
              "          quickchartButtonEl.style.display =\n",
              "            google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "        })();\n",
              "      </script>\n",
              "    </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "summary": "{\n  \"name\": \"display(non_cancer_df_row\",\n  \"rows\": 5,\n  \"fields\": [\n    {\n      \"column\": \"texts\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 5,\n        \"samples\": [\n          \"<ID:31164412>\\nTitle: T-Cell Deletion of MyD88 Connects IL17 and IkappaBzeta to RAS Oncogenesis.\\nAbstract: Cancer development requires a favorable tissue microenvironment. By deleting Myd88 in keratinocytes or specific bone marrow subpopulations in oncogenic RAS-mediated skin carcinogenesis, we show that IL17 from infiltrating T cells and IkappaBzeta signaling in keratinocytes are essential to produce a permissive microenvironment and tumor formation. Both normal and RAS-transformed keratinocytes respond to tumor promoters by activating canonical NF-kappaB and IkappaBzeta signaling, releasing specific cytokines and chemokines that attract Th17 cells through MyD88-dependent signaling in T cells. The release of IL17 into the microenvironment elevates IkappaBzeta in normal and RAS-transformed keratinocytes. Activation of IkappaBzeta signaling is required for the expression of specific promoting factors induced by IL17 in normal keratinocytes and constitutively expressed in RAS-initiated keratinocytes. Deletion of Nfkbiz in keratinocytes impairs RAS-mediated benign tumor formation. Transcriptional profiling and gene set enrichment analysis of IkappaBzeta-deficient RAS-initiated keratinocytes indicate that IkappaBzeta signaling is common for RAS transformation of multiple epithelial cancers. Probing The Cancer Genome Atlas datasets using this transcriptional profile indicates that reduction of IkappaBzeta signaling during cancer progression associates with poor prognosis in RAS-driven human cancers. IMPLICATIONS: The paradox that elevation of IkappaBzeta and stimulation of IkappaBzeta signaling through tumor extrinsic factors is required for RAS-mediated benign tumor formation while relative IkappaBzeta expression is reduced in advanced cancers with poor prognosis implies that tumor cells switch from microenvironmental dependency early in carcinogenesis to cell-autonomous pathways during cancer progression.\",\n          \"<ID:30897768>\\nTitle: Breast Cancer and miR-SNPs: The Importance of miR Germ-Line Genetics.\\nAbstract: Recent studies in cancer diagnostics have identified microRNAs (miRNAs) as promising cancer biomarkers. Single nucleotide polymorphisms (SNPs) in miRNA binding sites, seed regions, and coding sequences can help predict breast cancer risk, aggressiveness, response to stimuli, and prognosis. This review also documents significant known miR-SNPs in miRNA biogenesis genes and their effects on gene regulation in breast cancer, taking into account the genetic background and ethnicity of the sampled populations. When applicable, miR-SNPs are evaluated in the context of other patient factors, including mutations, hormonal status, and demographics. Given the power of miR-SNPs to predict patient cancer risk, prognosis, and outcomes, further study of miR-SNPs is warranted to improve efforts towards personalized medicine.\",\n          \"<ID:31094905>\\nTitle: MYCN Amplified Relapse Following Resolution of MYCN Nonamplified 4S Neuroblastoma With Placental Involvement: A Case Report and Review of the Literature.\\nAbstract: Congenital neuroblastoma with placental involvement is exceptionally rare, but mortality is high. Detailed examination of placenta including MYCN amplification and segmental chromosomal aberrations should be performed in all suspected cases, as it is noninvasive and readily available. Maternal dissemination has not been reported. In this manuscript, we describe an infant with placental diagnosis of MYCN nonamplified congenital neuroblastoma. This is the first report of a recurrence of congenital 4S neuroblastoma following resolution in which MYCN amplification is only detected in the recurrence. Germline sequencing using a large comprehensive cancer panel did not reveal variants in candidate cancer predisposition genes.\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"filename\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 5,\n        \"samples\": [\n          \"31164412\",\n          \"30897768\",\n          \"31094905\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"labels\",\n      \"properties\": {\n        \"dtype\": \"boolean\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          true\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "                                               texts  filename  labels\n",
              "0  <ID:25486933>\\nTitle: Strategy for identifying...  25486933   False\n",
              "1  <ID:28699658>\\nTitle: Telomere length and soma...  28699658   False\n",
              "2  <ID:31026806>\\nTitle: Transcription Factor STA...  31026806   False\n",
              "3  <ID:30255984>\\nTitle: Genetics, genomics, and ...  30255984   False\n",
              "4  <ID:26706013>\\nTitle: Genotype/Phenotype Corre...  26706013   False"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-91feb8f3-515f-43e9-8cf3-caed09224671\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>texts</th>\n",
              "      <th>filename</th>\n",
              "      <th>labels</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>&lt;ID:25486933&gt;\\nTitle: Strategy for identifying...</td>\n",
              "      <td>25486933</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>&lt;ID:28699658&gt;\\nTitle: Telomere length and soma...</td>\n",
              "      <td>28699658</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>&lt;ID:31026806&gt;\\nTitle: Transcription Factor STA...</td>\n",
              "      <td>31026806</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>&lt;ID:30255984&gt;\\nTitle: Genetics, genomics, and ...</td>\n",
              "      <td>30255984</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>&lt;ID:26706013&gt;\\nTitle: Genotype/Phenotype Corre...</td>\n",
              "      <td>26706013</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-91feb8f3-515f-43e9-8cf3-caed09224671')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-91feb8f3-515f-43e9-8cf3-caed09224671 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-91feb8f3-515f-43e9-8cf3-caed09224671');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    <div id=\"df-f1636f1b-40df-4b78-8657-bdf1fafb2611\">\n",
              "      <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-f1636f1b-40df-4b78-8657-bdf1fafb2611')\"\n",
              "                title=\"Suggest charts\"\n",
              "                style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "      </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "      <script>\n",
              "        async function quickchart(key) {\n",
              "          const quickchartButtonEl =\n",
              "            document.querySelector('#' + key + ' button');\n",
              "          quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "          quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "          try {\n",
              "            const charts = await google.colab.kernel.invokeFunction(\n",
              "                'suggestCharts', [key], {});\n",
              "          } catch (error) {\n",
              "            console.error('Error during call to suggestCharts:', error);\n",
              "          }\n",
              "          quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "          quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "        }\n",
              "        (() => {\n",
              "          let quickchartButtonEl =\n",
              "            document.querySelector('#df-f1636f1b-40df-4b78-8657-bdf1fafb2611 button');\n",
              "          quickchartButtonEl.style.display =\n",
              "            google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "        })();\n",
              "      </script>\n",
              "    </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "summary": "{\n  \"name\": \"display(non_cancer_df_row\",\n  \"rows\": 5,\n  \"fields\": [\n    {\n      \"column\": \"texts\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 5,\n        \"samples\": [\n          \"<ID:28699658>\\nTitle: Telomere length and somatic mutations in correlation with response to immunosuppressive treatment in aplastic anaemia.\\nAbstract: We investigated the frequencies of cytogenetic aberrations and somatic mutations of prognostic relevance in 393 patients with aplastic anaemia (AA). Clonality was determined by G-banding/fluorescence in situ hybridization (FISH) (n = 245), and targeted capture sequencing was performed for 88 haematopoiesis-related genes (n = 70). The telomere length (TL) of bone marrow nucleated cells was measured at the single cell level by FISH (n = 135). Eighteen (4 6%) patients showed disease progression, and monosomy 7 (50 0%) was the most predominant cytogenetic evolution at disease transformation. One third of patients (32 9%) presented at least 1 mutation; the most frequently mutated genes were NOTCH1, NF1, SCRIB, BCOR and DNMT3A. The patient group with clonal changes (30 7%) showed an adverse response to immunosuppressive treatment (IST), compared to the non-clonal group, but this finding did not show statistical significance. The TL of AA patients was significantly shorter than normal control and patients with clonal changes showed significantly shorter TLs. Patients with TL>5 9 showed a higher response rate to IST (P = 0 048). In conclusion, the patients with clonal changes or TL attrition showed a poor response to IST. Shorter TL can be used not only as a biomarker, but also as a predictive marker for treatment response to IST.\",\n          \"<ID:26706013>\\nTitle: Genotype/Phenotype Correlations in Tuberous Sclerosis Complex.\\nAbstract: Tuberous sclerosis complex (TSC) is an autosomal dominant disorder characterized by the development of widespread hamartomatous lesions in various organs, including brain, skin, kidneys, heart, and eyes. Central nervous system is almost invariably involved, with up to 85% of patients presenting with epilepsy, and at least half of patients having intellectual disability or other neuropsychiatric disorders including autism spectrum disorder. TSC is caused by the mutation in one of the 2 genes TSC1, at 9q34, and TSC2, at 16p13.3. They respectively encode for hamartin and tuberin, which form an intracellular complex inhibiting the mammalian target of rapamycin. Mammalian target of rapamycin overactivation following the genetic defect determines the cell growth and proliferation responsible for TSC-related lesions, as well as the alterations in neuronal excitability and synaptogenesis leading to epilepsy and neuropsychiatric disorders. A causative mutation for the disorder is identified in about 85% of patients with a clinical diagnosis of TSC. Mosaicism and technology limits likely explain most of the no mutation identified cases. This review confirms that patients with TSC2 mutations considered as a group usually present a more severe phenotype, characterized by higher number of tubers, earlier age at seizure onset and higher prevalence of intellectual disability. However, the clinical phenotype of the disease presents a high variability, thus making the prediction of the phenotype on an individual basis still challenging. The increasing application of new molecular techniques to subjects with TSC has the potential to significantly reduce the rate of patients with no mutation demonstrated and to identify an increasing higher number of mutations. This would hopefully allow a better characterization of higher risk mutations, which might help clinicians to plan individualized surveillance plans. Furthermore, the increasing availability of disease registries to collect clinical and genetics data of patients help to define more valid and clinically oriented genotype or phenotype correlations.\",\n          \"<ID:31026806>\\nTitle: Transcription Factor STAT3 Serves as a Negative Regulator Controlling IgE Class Switching in Mice.\\nAbstract: A mutation in STAT3 has been linked to the incidence of autosomal dominant hyper IgE syndrome, a disease characterized by elevated serum IgE Ab. However, how this genetic mutation leads to the phenotype has not been fully understood. We investigated the specific role of STAT3 in the germinal center (GC) B cells and plasma cells for IgE class switching. Through the use of STAT3 conditional knockout (cKO) mice in a Th2-type immunization model, we demonstrated that CD2-Cre-driven STAT3 cKO mice showed elevated IgE and decreased IgG1 in the serum and a reduction in GC formation. Within the GC, IgG1 + GC B cells were decreased, whereas IgE+ GC B cells were more prevalent. Additionally, these mice exhibited reduced IgG1 and elevated IgE populations of Ab-producing plasma cells. Subsequent experiments using a CD19-Cre cKO mouse established this effect to be B cell-intrinsic. Transcription factors critical for GC and plasma cell differentiation, including Bcl-6 and Aicda, were shown to function as downstream signals of STAT3 regulation. Chromatin immunoprecipitation sequencing analysis revealed that many genes, including Bcl3 and Crtc2, were among the direct STAT3 regulated targets. Mice with STAT3 deficiency in B cells also demonstrated an increase in lung inflammation when used in an asthma-like disease model. This model suggests a negative role for STAT3 in regulating class switching of the GC B cells from the IgG1 to the IgE producing state, which may serve as a therapeutic target for treatment of autosomal dominant hyper IgE syndrome and other immune disorders.\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"filename\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 5,\n        \"samples\": [\n          \"28699658\",\n          \"26706013\",\n          \"31026806\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"labels\",\n      \"properties\": {\n        \"dtype\": \"boolean\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          false\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {}
        }
      ],
      "source": [
        "zip_path=\"Dataset.zip\"\n",
        "cancer_df_row = load_text_files_from_zip_to_dataframe(zip_path, 'Dataset/Cancer')\n",
        "cancer_df_row['labels'] = True\n",
        "\n",
        "non_cancer_df_row = load_text_files_from_zip_to_dataframe(zip_path, 'Dataset/Non-Cancer')\n",
        "non_cancer_df_row['labels'] = False\n",
        "\n",
        "display(cancer_df_row.head())\n",
        "display(non_cancer_df_row.head())"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "E1XjUmvui6YZ"
      },
      "source": [
        "# concat cancer_df_row, non_cancer_df_row"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 423
        },
        "id": "TRJ_pSKDi5x8",
        "outputId": "0addeed8-668e-4038-ed1c-809c13c32de5"
      },
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "                                                 texts  filename  labels\n",
              "0    <ID:31055803>\\nTitle: [Analysis of age-specifi...  31055803    True\n",
              "1    <ID:31164412>\\nTitle: T-Cell Deletion of MyD88...  31164412    True\n",
              "2    <ID:31094905>\\nTitle: MYCN Amplified Relapse F...  31094905    True\n",
              "3    <ID:31498304>\\nTitle: In Vivo Inhibition of Mi...  31498304    True\n",
              "4    <ID:30897768>\\nTitle: Breast Cancer and miR-SN...  30897768    True\n",
              "..                                                 ...       ...     ...\n",
              "995  <ID:26095439>\\nTitle: Urinary 11beta-PGF2alpha...  26095439   False\n",
              "996  <ID:24850616>\\nTitle: A limited form of proteu...  24850616   False\n",
              "997  <ID:24402730>\\nTitle: Benign mast cell hyperpl...  24402730   False\n",
              "998  <ID:26513044>\\nTitle: Nevus anemicus associate...  26513044   False\n",
              "999  <ID:24688314>\\nTitle: Comparison of methods fo...  24688314   False\n",
              "\n",
              "[1000 rows x 3 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-6ca602f8-4888-4090-b1c4-aee98b24682e\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>texts</th>\n",
              "      <th>filename</th>\n",
              "      <th>labels</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>&lt;ID:31055803&gt;\\nTitle: [Analysis of age-specifi...</td>\n",
              "      <td>31055803</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>&lt;ID:31164412&gt;\\nTitle: T-Cell Deletion of MyD88...</td>\n",
              "      <td>31164412</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>&lt;ID:31094905&gt;\\nTitle: MYCN Amplified Relapse F...</td>\n",
              "      <td>31094905</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>&lt;ID:31498304&gt;\\nTitle: In Vivo Inhibition of Mi...</td>\n",
              "      <td>31498304</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>&lt;ID:30897768&gt;\\nTitle: Breast Cancer and miR-SN...</td>\n",
              "      <td>30897768</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>995</th>\n",
              "      <td>&lt;ID:26095439&gt;\\nTitle: Urinary 11beta-PGF2alpha...</td>\n",
              "      <td>26095439</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>996</th>\n",
              "      <td>&lt;ID:24850616&gt;\\nTitle: A limited form of proteu...</td>\n",
              "      <td>24850616</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>997</th>\n",
              "      <td>&lt;ID:24402730&gt;\\nTitle: Benign mast cell hyperpl...</td>\n",
              "      <td>24402730</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>998</th>\n",
              "      <td>&lt;ID:26513044&gt;\\nTitle: Nevus anemicus associate...</td>\n",
              "      <td>26513044</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>999</th>\n",
              "      <td>&lt;ID:24688314&gt;\\nTitle: Comparison of methods fo...</td>\n",
              "      <td>24688314</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>1000 rows × 3 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-6ca602f8-4888-4090-b1c4-aee98b24682e')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-6ca602f8-4888-4090-b1c4-aee98b24682e button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-6ca602f8-4888-4090-b1c4-aee98b24682e');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    <div id=\"df-3df90d18-2635-43ca-967d-4c4b9a141151\">\n",
              "      <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-3df90d18-2635-43ca-967d-4c4b9a141151')\"\n",
              "                title=\"Suggest charts\"\n",
              "                style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "      </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "      <script>\n",
              "        async function quickchart(key) {\n",
              "          const quickchartButtonEl =\n",
              "            document.querySelector('#' + key + ' button');\n",
              "          quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "          quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "          try {\n",
              "            const charts = await google.colab.kernel.invokeFunction(\n",
              "                'suggestCharts', [key], {});\n",
              "          } catch (error) {\n",
              "            console.error('Error during call to suggestCharts:', error);\n",
              "          }\n",
              "          quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "          quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "        }\n",
              "        (() => {\n",
              "          let quickchartButtonEl =\n",
              "            document.querySelector('#df-3df90d18-2635-43ca-967d-4c4b9a141151 button');\n",
              "          quickchartButtonEl.style.display =\n",
              "            google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "        })();\n",
              "      </script>\n",
              "    </div>\n",
              "\n",
              "  <div id=\"id_f4797371-90a8-4ac0-b5f5-fa1ee08eae4f\">\n",
              "    <style>\n",
              "      .colab-df-generate {\n",
              "        background-color: #E8F0FE;\n",
              "        border: none;\n",
              "        border-radius: 50%;\n",
              "        cursor: pointer;\n",
              "        display: none;\n",
              "        fill: #1967D2;\n",
              "        height: 32px;\n",
              "        padding: 0 0 0 0;\n",
              "        width: 32px;\n",
              "      }\n",
              "\n",
              "      .colab-df-generate:hover {\n",
              "        background-color: #E2EBFA;\n",
              "        box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "        fill: #174EA6;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate {\n",
              "        background-color: #3B4455;\n",
              "        fill: #D2E3FC;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate:hover {\n",
              "        background-color: #434B5C;\n",
              "        box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "        filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "        fill: #FFFFFF;\n",
              "      }\n",
              "    </style>\n",
              "    <button class=\"colab-df-generate\" onclick=\"generateWithVariable('data_df')\"\n",
              "            title=\"Generate code using this dataframe.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M7,19H8.4L18.45,9,17,7.55,7,17.6ZM5,21V16.75L18.45,3.32a2,2,0,0,1,2.83,0l1.4,1.43a1.91,1.91,0,0,1,.58,1.4,1.91,1.91,0,0,1-.58,1.4L9.25,21ZM18.45,9,17,7.55Zm-12,3A5.31,5.31,0,0,0,4.9,8.1,5.31,5.31,0,0,0,1,6.5,5.31,5.31,0,0,0,4.9,4.9,5.31,5.31,0,0,0,6.5,1,5.31,5.31,0,0,0,8.1,4.9,5.31,5.31,0,0,0,12,6.5,5.46,5.46,0,0,0,6.5,12Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "    <script>\n",
              "      (() => {\n",
              "      const buttonEl =\n",
              "        document.querySelector('#id_f4797371-90a8-4ac0-b5f5-fa1ee08eae4f button.colab-df-generate');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      buttonEl.onclick = () => {\n",
              "        google.colab.notebook.generateWithVariable('data_df');\n",
              "      }\n",
              "      })();\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "data_df",
              "summary": "{\n  \"name\": \"data_df\",\n  \"rows\": 1000,\n  \"fields\": [\n    {\n      \"column\": \"texts\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 1000,\n        \"samples\": [\n          \"<ID:25899109>\\nTitle: Anaphylaxis with clonal mast cells in normal looking skin - a new entity?\\nAbstract: BACKGROUND: Patients with elevated basal tryptase (sBT) >15 mug/l and anaphylaxis may have an underlying mastocytosis. A monoclonal mast cell activation syndrome with aberrant mast cells (MC) at extracutaneous sites has been described in patients with severe hypotension or anaphylaxis. METHODS: As MC in patients with elevated sBT might be altered in the skin as well, we studied MC in normal neck skin in anaphylaxis and urticaria patients with elevated sBT. RESULTS: A mean of 93.1 (SD 19.1) MC/mm2 was counted in normal neck skin in 14 patients with anaphylaxis, 84.0 (SD 13.6) in seven patients with urticaria, 142.0 (SD 24.0) in two patients with eczema, 124.4 (SD 43.2) in five patients with systemic mastocytosis (SM) in comparison with autopsy skin (39.1 MC/mm2, SD 12.4). In five of 14 (35.7%) of the anaphylaxis and three of five (60%) SM patients more than 25% of MC were spindle shaped and expressed CD25 antigen. CONCLUSIONS: We could show for the first time that the normal skin can harbour clonal MC in anaphylaxis patients. Analogous to the criteria for mastocytosis, we suggest a skin score criteria including an elevated number of MC, spindle shape, CD25 expression, c-Kit mutation and sBT values >20 mug/l. In patients with anaphylaxis and elevated sBT, skin should be biopsied and, as with the approach for mastocytosis diagnosis in the bone marrow, MC should be analysed for their number, clonality and c-Kit mutation. This approach should be confirmed in further studies. Patients with aberrant skin MC should be handled as mastocytosis patients.\",\n          \"<ID:25476696>\\nTitle: Intestinal epithelial MyD88 is a sensor switching host metabolism towards obesity according to nutritional status.\\nAbstract: Obesity is associated with a cluster of metabolic disorders, low-grade inflammation and altered gut microbiota. Whether host metabolism is controlled by intestinal innate immune system and the gut microbiota is unknown. Here we report that inducible intestinal epithelial cell-specific deletion of MyD88 partially protects against diet-induced obesity, diabetes and inflammation. This is associated with increased energy expenditure, an improved glucose homeostasis, reduced hepatic steatosis, fat mass and inflammation. Protection is transferred following gut microbiota transplantation to germ-free recipients. We also demonstrate that intestinal epithelial MyD88 deletion increases anti-inflammatory endocannabinoids, restores antimicrobial peptides production and increases intestinal regulatory T cells during diet-induced obesity. Targeting MyD88 after the onset of obesity reduces fat mass and inflammation. Our work thus identifies intestinal epithelial MyD88 as a sensor changing host metabolism according to the nutritional status and we show that targeting intestinal epithelial MyD88 constitutes a putative therapeutic target for obesity and related disorders.\",\n          \"<ID:30545997>\\nTitle: Systemic Mastocytosis, Version 2.2019, NCCN Clinical Practice Guidelines in Oncology.\\nAbstract: Mastocytosis is a group of heterogeneous disorders resulting from the clonal proliferation of abnormal mast cells and their accumulation in the skin and/or in various extracutaneous organs. Systemic mastocytosis is the most common form of mastocytosis diagnosed in adults, characterized by mast cell infiltration of one or more extracutaneous organs (with or without skin involvement). The identification of KIT D816V mutation and the emergence of novel targeted therapies have significantly improved the diagnosis and treatment of systemic mastocytosis. However, certain aspects of clinical care, particularly the diagnosis, assessment, and management of mediator-related symptoms continue to present challenges. This manuscript discusses the recommendations outlined in the NCCN Guidelines for the diagnosis and management of patients with systemic mastocytosis.\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"filename\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 1000,\n        \"samples\": [\n          \"25899109\",\n          \"25476696\",\n          \"30545997\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"labels\",\n      \"properties\": {\n        \"dtype\": \"boolean\",\n        \"num_unique_values\": 2,\n        \"samples\": [\n          false,\n          true\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {}
        }
      ],
      "source": [
        "data_df = pd.concat([cancer_df_row, non_cancer_df_row], ignore_index=True)\n",
        "display(data_df)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "1VqfwpqkjQWc"
      },
      "source": [
        "# split cancer data into train,test,val"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "metadata": {
        "id": "2g-PQ7UdjY0Z"
      },
      "outputs": [],
      "source": [
        "from sklearn.model_selection import train_test_split\n",
        "\n",
        "def split_train_test_val(df, train_size=0.6, val_size=0.2, test_size=0.2, random_state=42):\n",
        "    \"\"\"\n",
        "    Splits a DataFrame into train, test, and validation sets with specified ratios.\n",
        "    \"\"\"\n",
        "    if not isinstance(df, pd.DataFrame):\n",
        "        raise TypeError(\"Input must be a Pandas DataFrame.\")\n",
        "\n",
        "    if sum([train_size, test_size, val_size]) != 1.0:\n",
        "        raise ValueError(\"Train, test, and validation sizes must sum to 1.\")\n",
        "\n",
        "    # First split into training and remaining data\n",
        "    train_df, remaining_df = train_test_split(df, train_size=train_size, random_state=random_state)\n",
        "\n",
        "    # Calculate the proportion for test and validation sets from the remaining data\n",
        "    remaining_proportion = 1 - train_size\n",
        "    test_proportion = test_size / remaining_proportion\n",
        "    val_proportion = val_size / remaining_proportion\n",
        "\n",
        "    # Split the remaining data into test and validation sets\n",
        "    test_df, val_df = train_test_split(remaining_df, test_size=test_proportion, random_state=random_state)\n",
        "\n",
        "\n",
        "    return train_df, test_df, val_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 13,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 722
        },
        "id": "vLRUNdUcjeSi",
        "outputId": "b57c7036-ca83-42ab-c083-52322d72ccc7"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Cancer Train DataFrame:\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "         ID                                              texts  labels\n",
              "0  31186051  <ID:31186051>\\nTitle: Diverse regulatory manne...    True\n",
              "1  31681608  <ID:31681608>\\nTitle: Epigenetic Targets in Sy...    True\n",
              "2  27558955  <ID:27558955>\\nTitle: Molecular hydrogen suppr...   False\n",
              "3  29930110  <ID:29930110>\\nTitle: Analysis of shared herit...   False\n",
              "4  36928589  <ID:36928589>\\nTitle: Synthetic multiantigen M...   False"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-50ae2d31-2792-44d4-bef0-8336c96e2af9\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>ID</th>\n",
              "      <th>texts</th>\n",
              "      <th>labels</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>31186051</td>\n",
              "      <td>&lt;ID:31186051&gt;\\nTitle: Diverse regulatory manne...</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>31681608</td>\n",
              "      <td>&lt;ID:31681608&gt;\\nTitle: Epigenetic Targets in Sy...</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>27558955</td>\n",
              "      <td>&lt;ID:27558955&gt;\\nTitle: Molecular hydrogen suppr...</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>29930110</td>\n",
              "      <td>&lt;ID:29930110&gt;\\nTitle: Analysis of shared herit...</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>36928589</td>\n",
              "      <td>&lt;ID:36928589&gt;\\nTitle: Synthetic multiantigen M...</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-50ae2d31-2792-44d4-bef0-8336c96e2af9')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-50ae2d31-2792-44d4-bef0-8336c96e2af9 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-50ae2d31-2792-44d4-bef0-8336c96e2af9');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    <div id=\"df-a1cc4ab0-d1f1-4578-ae71-f15aa707112e\">\n",
              "      <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-a1cc4ab0-d1f1-4578-ae71-f15aa707112e')\"\n",
              "                title=\"Suggest charts\"\n",
              "                style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "      </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "      <script>\n",
              "        async function quickchart(key) {\n",
              "          const quickchartButtonEl =\n",
              "            document.querySelector('#' + key + ' button');\n",
              "          quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "          quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "          try {\n",
              "            const charts = await google.colab.kernel.invokeFunction(\n",
              "                'suggestCharts', [key], {});\n",
              "          } catch (error) {\n",
              "            console.error('Error during call to suggestCharts:', error);\n",
              "          }\n",
              "          quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "          quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "        }\n",
              "        (() => {\n",
              "          let quickchartButtonEl =\n",
              "            document.querySelector('#df-a1cc4ab0-d1f1-4578-ae71-f15aa707112e button');\n",
              "          quickchartButtonEl.style.display =\n",
              "            google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "        })();\n",
              "      </script>\n",
              "    </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "summary": "{\n  \"name\": \"print(f\\\"Validation shape: {val_df\",\n  \"rows\": 5,\n  \"fields\": [\n    {\n      \"column\": \"ID\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 5,\n        \"samples\": [\n          \"31681608\",\n          \"36928589\",\n          \"27558955\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"texts\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 5,\n        \"samples\": [\n          \"<ID:31681608>\\nTitle: Epigenetic Targets in Synovial Sarcoma: A Mini-Review.\\nAbstract: Synovial Sarcomas (SS) are a type of Soft Tissue Sarcoma (STS) and represent 8-10% of all STS cases. Although SS can arise at any age, it typically affects younger individuals aged 15-35 and is therefore part of both pediatric and adult clinical practices. SS occurs primarily in the limbs, often near joints, but can present anywhere. It is characterized by the recurrent pathognomonic chromosomal translocation t(X;18)(p11.2;q11.2) that most frequently fuses SSX1 or SSX2 genes with SS18. This leads to the expression of the SS18-SSX fusion protein, which causes disturbances in several interacting multiprotein complexes such as the SWItch/Sucrose Non-Fermentable (SWI/SNF) complex, also known as the BAF complex and the Polycomb Repressive Complex 1 and 2 (PRC1 and PRC2). Furthermore, this promotes widespread epigenetic rewiring, leading to aberrant gene expression that drives the pathogenesis of SS. Good prognoses are characterized predominantly by small tumor size and young patient age. Whereas, high tumor grade and an increased genomic complexity of the tumor constitute poor prognostic factors. The current therapeutic strategy relies on chemotherapy and radiotherapy, the latter of which can lead to chronic side effects for pediatric patients. We will focus on the known roles of SWI/SNF, PRC1, and PRC2 as the main effectors of the SS18-SSX-mediated genome modifications and we present existing biological rationale for potential therapeutic targets and treatment strategies.\",\n          \"<ID:36928589>\\nTitle: Synthetic multiantigen MVA vaccine COH04S1 and variant-specific derivatives protect Syrian hamsters from SARS-CoV-2 Omicron subvariants.\\nAbstract: Emerging SARS-CoV-2 Omicron subvariants continue to disrupt COVID-19 vaccine efficacy through multiple immune mechanisms including neutralizing antibody evasion. We developed COH04S1, a synthetic modified vaccinia Ankara vector that co-expresses Wuhan-Hu-1-based spike and nucleocapsid antigens. COH04S1 demonstrated efficacy against ancestral virus and Beta and Delta variants in animal models and was safe and immunogenic in a Phase 1 clinical trial. Here, we report efficacy of COH04S1 and analogous Omicron BA.1- and Beta-specific vaccines to protect Syrian hamsters from Omicron subvariants. Despite eliciting strain-specific antibody responses, all three vaccines protect hamsters from weight loss, lower respiratory tract infection, and lung pathology following challenge with Omicron BA.1 or BA.2.12.1. While the BA.1-specifc vaccine affords consistently improved efficacy compared to COH04S1 to protect against homologous challenge with BA.1, all three vaccines confer similar protection against heterologous challenge with BA.2.12.1. These results demonstrate efficacy of COH04S1 and variant-specific derivatives to confer cross-protective immunity against SARS-CoV-2 Omicron subvariants.\",\n          \"<ID:27558955>\\nTitle: Molecular hydrogen suppresses activated Wnt/beta-catenin signaling.\\nAbstract: Molecular hydrogen (H2) is effective for many diseases. However, molecular bases of H2 have not been fully elucidated. Cumulative evidence indicates that H2 acts as a gaseous signal modulator. We found that H2 suppresses activated Wnt/beta-catenin signaling by promoting phosphorylation and degradation omicronf beta-catenin. Either complete inhibition of GSK3 or mutations at CK1- and GSK3-phosphorylation sites of beta-catenin abolished the suppressive effect of H2. H2 did not increase GSK3-mediated phosphorylation of glycogen synthase, indicating that H2 has no direct effect on GSK3 itself. Knock-down of adenomatous polyposis coli (APC) or Axin1, which form the beta-catenin degradation complex, minimized the suppressive effect of H2 on beta-catenin accumulation. Accordingly, the effect of H2 requires CK1/GSK3-phosphorylation sites of beta-catenin, as well as the beta-catenin degradation complex comprised of CK1, GSK3, APC, and Axin1. We additionally found that H2 reduces the activation of Wnt/beta-catenin signaling in human osteoarthritis chondrocytes. Oral intake of H2 water tended to ameliorate cartilage degradation in a surgery-induced rat osteoarthritis model through attenuating beta-catenin accumulation. We first demonstrate that H2 suppresses abnormally activated Wnt/beta-catenin signaling, which accounts for the protective roles of H2 in a fraction of diseases.\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"labels\",\n      \"properties\": {\n        \"dtype\": \"boolean\",\n        \"num_unique_values\": 2,\n        \"samples\": [\n          false,\n          true\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Train shape: (600, 3)\n",
            "\n",
            "Cancer Test DataFrame:\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "         ID                                              texts  labels\n",
              "0  29728436  <ID:29728436>\\nTitle: Splenic cyst and its man...   False\n",
              "1  24906339  <ID:24906339>\\nTitle: Tracking histone variant...   False\n",
              "2  31602389  <ID:31602389>\\nTitle: Molecular Approaches to ...    True\n",
              "3  31244296  <ID:31244296>\\nTitle: Identification of FLT3 a...    True\n",
              "4  36610490  <ID:36610490>\\nTitle: Phase I Study: Safety an...    True"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-6ff3ff39-7aa1-437b-b5b2-c400a57502f3\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>ID</th>\n",
              "      <th>texts</th>\n",
              "      <th>labels</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>29728436</td>\n",
              "      <td>&lt;ID:29728436&gt;\\nTitle: Splenic cyst and its man...</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>24906339</td>\n",
              "      <td>&lt;ID:24906339&gt;\\nTitle: Tracking histone variant...</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>31602389</td>\n",
              "      <td>&lt;ID:31602389&gt;\\nTitle: Molecular Approaches to ...</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>31244296</td>\n",
              "      <td>&lt;ID:31244296&gt;\\nTitle: Identification of FLT3 a...</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>36610490</td>\n",
              "      <td>&lt;ID:36610490&gt;\\nTitle: Phase I Study: Safety an...</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-6ff3ff39-7aa1-437b-b5b2-c400a57502f3')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-6ff3ff39-7aa1-437b-b5b2-c400a57502f3 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-6ff3ff39-7aa1-437b-b5b2-c400a57502f3');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    <div id=\"df-310266ef-81c9-4a17-84f5-7b4559dcf71c\">\n",
              "      <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-310266ef-81c9-4a17-84f5-7b4559dcf71c')\"\n",
              "                title=\"Suggest charts\"\n",
              "                style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "      </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "      <script>\n",
              "        async function quickchart(key) {\n",
              "          const quickchartButtonEl =\n",
              "            document.querySelector('#' + key + ' button');\n",
              "          quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "          quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "          try {\n",
              "            const charts = await google.colab.kernel.invokeFunction(\n",
              "                'suggestCharts', [key], {});\n",
              "          } catch (error) {\n",
              "            console.error('Error during call to suggestCharts:', error);\n",
              "          }\n",
              "          quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "          quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "        }\n",
              "        (() => {\n",
              "          let quickchartButtonEl =\n",
              "            document.querySelector('#df-310266ef-81c9-4a17-84f5-7b4559dcf71c button');\n",
              "          quickchartButtonEl.style.display =\n",
              "            google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "        })();\n",
              "      </script>\n",
              "    </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "summary": "{\n  \"name\": \"print(f\\\"Validation shape: {val_df\",\n  \"rows\": 5,\n  \"fields\": [\n    {\n      \"column\": \"ID\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 5,\n        \"samples\": [\n          \"24906339\",\n          \"36610490\",\n          \"31602389\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"texts\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 5,\n        \"samples\": [\n          \"<ID:24906339>\\nTitle: Tracking histone variant nucleosomes across the human cell cycle using biophysical, biochemical, and cytological analyses.\\nAbstract: Histone variants such as H3.3, macroH2A, H2A.Z, and CENP-A are important epigenetic modifiers of the chromatin state in eukaryotic genomes. The centromeric histone H3 variant CENP-A/CENH3 epigenetically marks centromeres and is required for assembly of the kinetochore complex, a region of the chromosome that is responsible for proper genome segregation during mitosis. Several diverse techniques using biochemical, cell biology, and biophysical approaches have been utilized to study the nature of the CENP-A nucleosome across the cell cycle. In this chapter, we describe methods for CENP-A nucleosome purification and separation of CENP-A from other core histones using traditional SDS-PAGE and more resolving techniques such as Triton acid urea (TAU) and two-dimensional gels. We also discuss methods for observation of CENP-A on chromatin fibers using immunofluorescence. Finally, we provide a detailed description of analysis of chromatin structures using atomic force microscopy.\",\n          \"<ID:36610490>\\nTitle: Phase I Study: Safety and Efficacy of an Ex Vivo-Expanded Allogeneic Natural Killer Cell (MG4101) with Rituximab for Relapsed/Refractory B Cell Non-Hodgkin Lymphoma.\\nAbstract: The prognosis of non-Hodgkin lymphoma (NHL) remains poor, with an unmet need for novel therapies. MG4101, an ex vivo-expanded allogeneic natural killer (NK) cell, can enhance rituximab antibody-dependent cytotoxicity in relapsed/refractory (r/r) B cell non-Hodgkin lymphoma. This study assessed the safety and efficacy of MG4101 plus rituximab for patients with r/r NHL. Patients received escalating doses of i.v. MG4101 plus rituximab every 2 weeks. IL-2 was administered s.c. after MG4101 treatment. Fludarabine plus cyclophosphamide was administered i.v. before rituximab treatment in cycles 1, 3, and 5. A 3+3 design was used to determine the maximum tolerated dose (MTD) and maximum feasible dose. Assessments were performed over a 6-cycle period, with an extended maintenance period of up to 8 cycles. Nine patients received 3 different doses of MG4101 and rituximab. MTD could not be determined because of the absence of dose-limiting toxicity. Treatment-related adverse events, mostly grade 1 or 2, occurred in 89% of patients. Only 1 patient experienced grade 1 cytokine release syndrome. MG4101 persisted for at least 7 days in 7 patients. Four patients achieved a partial response and 1 patient attained a complete response, for an overall response rate of 55.6%. Two patients showed prolonged responses and low exhaustion marker levels in T cells. For allogeneic NK cell therapy, strategies including the use of the high-affinity hFc\\u03b3RIIIaV158 variant of the KIR B/x haplotype with lymphodepleting chemotherapy may be promising options for improving clinical efficacy in the antibody combination therapeutic setting as an off-the-shelf product. MG4101 plus rituximab presented a favorable safety profile and overall response rate in patients with r/r NHL.\",\n          \"<ID:31602389>\\nTitle: Molecular Approaches to Metastatic Colorectal Cancer: Better Diagnosis - Better Treatment?\\nAbstract: Molecular diagnostics of metastatic colorectal cancer (mCRC) are coming of age. Whereas for many years RAS mutations were the only relevant molecular alterations examined in mCRC, nowadays the BRAF, HER2, and microsatellite instability status of a given tumor is gaining importance, since it has not only prognostic relevance but opens new avenues for therapeutic approaches. This review summarizes the recent advances in molecular approaches to mCRC and their therapeutic consequences and discusses novel approaches for molecular diagnostics in mCRC.\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"labels\",\n      \"properties\": {\n        \"dtype\": \"boolean\",\n        \"num_unique_values\": 2,\n        \"samples\": [\n          true,\n          false\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Test shape: (200, 3)\n",
            "\n",
            "Cancer Validation DataFrame:\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "         ID                                              texts  labels\n",
              "0  38345107  <ID:38345107>\\nTitle: Similarities and differe...   False\n",
              "1  24777453  <ID:24777453>\\nTitle: Rare and low-frequency c...   False\n",
              "2  30973434  <ID:30973434>\\nTitle: The Effect of Inhaler Ar...    True\n",
              "3  31177124  <ID:31177124>\\nTitle: Paraoxonase 1 (PON1) Q19...    True\n",
              "4  31347028  <ID:31347028>\\nTitle: Clinical Evaluation of I...    True"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-b68c4b6f-d66f-427a-9f7a-673c40f10deb\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>ID</th>\n",
              "      <th>texts</th>\n",
              "      <th>labels</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>38345107</td>\n",
              "      <td>&lt;ID:38345107&gt;\\nTitle: Similarities and differe...</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>24777453</td>\n",
              "      <td>&lt;ID:24777453&gt;\\nTitle: Rare and low-frequency c...</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>30973434</td>\n",
              "      <td>&lt;ID:30973434&gt;\\nTitle: The Effect of Inhaler Ar...</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>31177124</td>\n",
              "      <td>&lt;ID:31177124&gt;\\nTitle: Paraoxonase 1 (PON1) Q19...</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>31347028</td>\n",
              "      <td>&lt;ID:31347028&gt;\\nTitle: Clinical Evaluation of I...</td>\n",
              "      <td>True</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-b68c4b6f-d66f-427a-9f7a-673c40f10deb')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-b68c4b6f-d66f-427a-9f7a-673c40f10deb button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-b68c4b6f-d66f-427a-9f7a-673c40f10deb');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    <div id=\"df-a5ed39dd-6fd2-4bf5-b287-3c6919c9f834\">\n",
              "      <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-a5ed39dd-6fd2-4bf5-b287-3c6919c9f834')\"\n",
              "                title=\"Suggest charts\"\n",
              "                style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "      </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "      <script>\n",
              "        async function quickchart(key) {\n",
              "          const quickchartButtonEl =\n",
              "            document.querySelector('#' + key + ' button');\n",
              "          quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "          quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "          try {\n",
              "            const charts = await google.colab.kernel.invokeFunction(\n",
              "                'suggestCharts', [key], {});\n",
              "          } catch (error) {\n",
              "            console.error('Error during call to suggestCharts:', error);\n",
              "          }\n",
              "          quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "          quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "        }\n",
              "        (() => {\n",
              "          let quickchartButtonEl =\n",
              "            document.querySelector('#df-a5ed39dd-6fd2-4bf5-b287-3c6919c9f834 button');\n",
              "          quickchartButtonEl.style.display =\n",
              "            google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "        })();\n",
              "      </script>\n",
              "    </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "summary": "{\n  \"name\": \"print(f\\\"Validation shape: {val_df\",\n  \"rows\": 5,\n  \"fields\": [\n    {\n      \"column\": \"ID\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 5,\n        \"samples\": [\n          \"24777453\",\n          \"31347028\",\n          \"30973434\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"texts\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 5,\n        \"samples\": [\n          \"<ID:24777453>\\nTitle: Rare and low-frequency coding variants in CXCR2 and other genes are associated with hematological traits.\\nAbstract: Hematological traits are important clinical parameters. To test the effects of rare and low-frequency coding variants on hematological traits, we analyzed hemoglobin concentration, hematocrit levels, white blood cell (WBC) counts and platelet counts in 31,340 individuals genotyped on an exome array. We identified several missense variants in CXCR2 associated with reduced WBC count (gene-based P = 2.6 x 10(-13)). In a separate family-based resequencing study, we identified a CXCR2 frameshift mutation in a pedigree with congenital neutropenia that abolished ligand-induced CXCR2 signal transduction and chemotaxis. We also identified missense or splice-site variants in key hematopoiesis regulators (EPO, TFR2, HBB, TUBB1 and SH2B3) associated with blood cell traits. Finally, we were able to detect associations between a rare somatic JAK2 mutation (encoding p.Val617Phe) and platelet count (P = 3.9 x 10(-22)) as well as hemoglobin concentration (P = 0.002), hematocrit levels (P = 9.5 x 10(-7)) and WBC count (P = 3.1 x 10(-5)). In conclusion, exome arrays complement genome-wide association studies in identifying new variants that contribute to complex human traits.\",\n          \"<ID:31347028>\\nTitle: Clinical Evaluation of IntelliPlex  KRAS G12/13 Mutation Kit for Detection of KRAS Mutations in Codon 12 and 13: A Novel Multiplex Approach.\\nAbstract: BACKGROUND: Colorectal cancer (CRC) is among the most frequently occurring cancers worldwide and its incidence is forecasted to increase. Testing for KRAS (Kirsten rat sarcoma viral oncogene homolog) mutations in colorectal tissue biopsy samples has become a crucial tool to guide therapeutic decisions for personalized treatment. OBJECTIVE: The objective of this study was to determine the diagnostic sensitivity and specificity of the IntelliPlex  KRAS G12/13 Mutation Kit using clinical specimens compared to Sanger sequencing as the reference method. METHODS: A total of 248 formalin-fixed paraffin-embedded (FFPE) tissue samples, with CRC tumors comprising more than 10% of the whole tissue sample, were included in the study and analyzed for specific KRAS mutations in codons 12 and 13. For samples with discordant results between Sanger sequencing and the IntelliPlex  KRAS G12/13 Mutation Kit, Pyrosequencing was utilized to resolve the KRAS mutational status. RESULTS: Sequencing determined 153 specimens as KRAS wild-type genotype while the IntelliPlex  KRAS G12/13 Mutation Kit confirmed 139 of the wild-type cases, resulting in a clinical specificity of 90.8% (95% confidence interval (CI) 85.12-94.91). All 95 specimens with a reported mutation in codons 12 or 13 of KRAS by sequencing were also reported as non-wild-type by the IntelliPlex  KRAS G12/13 Mutation Kit, resulting in a clinical sensitivity to detect KRAS mutations of 100% (95% CI 96.19-100). CONCLUSIONS: The IntelliPlex  KRAS G12/13 Mutation Kit demonstrates suitable specificity and sensitivity for use in clinical laboratories to determine the mutational status of KRAS codons 12 and 13.\",\n          \"<ID:30973434>\\nTitle: The Effect of Inhaler Aromatherapy on Invasive Pain, Procedure Adherence, Vital Signs, and Saturation During Port Catheterization in Oncology Patients.\\nAbstract: The study was conducted to evaluate the effect of inhaler aromatherapy on invasive pain, procedure adherence, vital signs, and saturation during port catheter insertion among patients diagnosed with cancer. The study was conducted in a nonrandomized controlled trial. Sixty patients including 30 patients in the intervention group and 30 patients in the control group, who were subjected with the same local anesthetic protocol, were included in the study. Aromatic mixture prepared by diluting orange, chamomile, and lavender oil in 70 mL distilled water was inhaled by the intervention group during the procedure. The data of the study were collected by using questionnaire, vital follow-up form, and visual analog scale. The patients in the intervention and control groups were similar in terms of sociodemographic and disease characteristics (P > .05). It was determined that inhaler aromatherapy applied to patients in the intervention group decreased pain experienced during the procedure and facilitated the procedure adherence (P < .05); however, it did not affect vital signs and saturation (P > .05). It can be recommended to administer inhaler aromatherapy with pharmacological therapies during catheterization procedure since it decreases invasive pain and facilitates the procedure adherence.\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"labels\",\n      \"properties\": {\n        \"dtype\": \"boolean\",\n        \"num_unique_values\": 2,\n        \"samples\": [\n          true,\n          false\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Validation shape: (200, 3)\n"
          ]
        }
      ],
      "source": [
        "train_df, test_df, val_df = split_train_test_val(data_df, train_size=0.6, val_size=0.2, test_size=0.2)\n",
        "\n",
        "\n",
        "## train_df reset index\n",
        "train_df.rename(columns={'filename': 'ID'}, inplace=True) # 1. Rename 'filename' column to 'ID'\n",
        "train_df = train_df.set_index('ID') # 2. Set 'ID' as the index (this will shift current index to the side)\n",
        "train_df = train_df.reset_index() # 3. Reset the index to remove the old index (the original integer index)\n",
        "\n",
        "## test_df reset index\n",
        "test_df.rename(columns={'filename': 'ID'}, inplace=True) # 1. Rename 'filename' column to 'ID'\n",
        "test_df = test_df.set_index('ID') # 2. Set 'ID' as the index (this will shift current index to the side)\n",
        "test_df = test_df.reset_index() # 3. Reset the index to remove the old index (the original integer index)\n",
        "\n",
        "## val_df reset index\n",
        "val_df.rename(columns={'filename': 'ID'}, inplace=True) # 1. Rename 'filename' column to 'ID'\n",
        "val_df = val_df.set_index('ID') # 2. Set 'ID' as the index (this will shift current index to the side)\n",
        "val_df = val_df.reset_index() # 3. Reset the index to remove the old index (the original integer index)\n",
        "\n",
        "print(\"Cancer Train DataFrame:\")\n",
        "display(train_df.head())\n",
        "print(f\"Train shape: {train_df.shape}\")\n",
        "\n",
        "print(\"\\nCancer Test DataFrame:\")\n",
        "display(test_df.head())\n",
        "print(f\"Test shape: {test_df.shape}\")\n",
        "\n",
        "print(\"\\nCancer Validation DataFrame:\")\n",
        "display(val_df.head())\n",
        "print(f\"Validation shape: {val_df.shape}\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "2XejbO4Y338f"
      },
      "source": [
        "preprocess text , add prompt and label"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 14,
      "metadata": {
        "id": "LjTZUdzmJsZh"
      },
      "outputs": [],
      "source": [
        "# Function to perform the text processing\n",
        "\n",
        "import re\n",
        "def process_text(row,isTestSetOrValSet):\n",
        "  # Compile the regex pattern (case-insensitive)\n",
        "  abstract_pattern = re.compile(r\"^\\s*Abstract:\\s*\", re.IGNORECASE | re.MULTILINE)  # Precompile regex\n",
        "\n",
        "  if not bool(abstract_pattern.search(row['texts'])):\n",
        "    text=\"Missing Abstract\"\n",
        "    return text\n",
        "\n",
        "  # Remove identifiers such as 'ID:', 'Title:', 'Abstract:'\n",
        "  # normalized_text = re.sub(r\"(ID:|Title:|Abstract:)\", \"\", row['texts'], flags=re.IGNORECASE)\n",
        "  normalized_text = re.sub(r\"<(?:ID:\\d+)>|Title:|Abstract:\", \"\", row['texts'], flags=re.IGNORECASE)\n",
        "\n",
        "  # Standardize common abbreviations\n",
        "  normalized_text = re.sub(r\"\\bet al\\.\\b\", \"and others\", normalized_text, flags=re.IGNORECASE)\n",
        "  normalized_text = re.sub(r\"\\bvol\\.\\b\", \"volume\", normalized_text, flags=re.IGNORECASE)\n",
        "\n",
        "  # Remove common punctuation marks, but leave periods at the end of sentences\n",
        "  normalized_text = re.sub(r\"[,;@#$%^&*()]\", \"\", normalized_text)\n",
        "\n",
        "  # Ensure consistent spacing, replaces multiple spaces with single ones, and adds space after periods.\n",
        "  normalized_text = \" \".join(normalized_text.split())\n",
        "  normalized_text = re.sub(r\"\\.(?=[A-Z])\", \". \", normalized_text)\n",
        "\n",
        "  # 2. create prompt ( I am adding this in dataset, alternativly it can be passed as function during training)\n",
        "  prompt=f\"You are a medical text classifier. Your task is to determine if the following text is related to cancer or not.\\\n",
        "          \\nLabel text only with 'True' or 'False'.\\\n",
        "          \\nIf text is related to cancer, Label as 'True' otherwise Label as 'False'.\\\n",
        "          \\nDo not provide any additional explanation.\\\n",
        "          \\n\\nText:\\n\"\n",
        "\n",
        "\n",
        "\n",
        "  text = prompt + normalized_text.strip() + \"\\n\\nLabel: \"\n",
        "\n",
        "\n",
        "\n",
        "  # 3. Append actual label in case train dataset\n",
        "  if not isTestSetOrValSet:\n",
        "    text = text + str(row['labels'])\n",
        "\n",
        "  return text"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 15,
      "metadata": {
        "id": "Y3aejUIi2BAr"
      },
      "outputs": [],
      "source": [
        "# Apply the processing function with the param isTestSetOrValSet\n",
        "train_df['text'] = train_df.apply(lambda row: process_text(row, isTestSetOrValSet=False), axis=1)\n",
        "train_df.drop(columns=['texts'], inplace=True) # Drop the original 'texts' column\n",
        "\n",
        "test_df['text'] = test_df.apply(lambda row: process_text(row, isTestSetOrValSet=True), axis=1)\n",
        "test_df.drop(columns=['texts'], inplace=True) # Drop the original 'texts' column\n",
        "\n",
        "val_df['text'] = val_df.apply(lambda row: process_text(row, isTestSetOrValSet=True), axis=1)\n",
        "val_df.drop(columns=['texts'], inplace=True) # Drop the original 'texts' column\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "gFO287fqPN0Y"
      },
      "source": [
        "Remove missing value"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 16,
      "metadata": {
        "id": "zfZwU0ijL2bg"
      },
      "outputs": [],
      "source": [
        "# Remove rows where 'text' column is \"Missing Abstract\"\n",
        "train_df = train_df[train_df['text'] != \"Missing Abstract\"]\n",
        "test_df = test_df[test_df['text'] != \"Missing Abstract\"]\n",
        "val_df = val_df[val_df['text'] != \"Missing Abstract\"]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 17,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 156
        },
        "id": "KLjrzKHa2WS6",
        "outputId": "0eb8531f-761e-40a8-b742-f57cbf83c1e3"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "\"You are a medical text classifier. Your task is to determine if the following text is related to cancer or not.          \\nLabel text only with 'True' or 'False'.          \\nIf text is related to cancer, Label as 'True' otherwise Label as 'False'.          \\nDo not provide any additional explanation.          \\n\\nText:\\nDiverse regulatory manners of human telomerase reverse transcriptase. Human telomerase reverse transcriptase hTERT is the core subunit of human telomerase and plays important roles in human cancers. Aberrant expression of hTERT is closely associated with tumorigenesis cancer cell stemness maintaining cell proliferation apoptosis inhibition senescence evasion and metastasis. The molecular basis of hTERT regulation is highly complicated and consists of various layers. A deep and full-scale comprehension of the regulatory mechanisms of hTERT is pivotal in understanding the pathogenesis and searching for therapeutic approaches. In this review we summarize the recent advances regarding the diverse regulatory mechanisms of hTERT including the transcriptional promoter mutation promoter region methylation and histone acetylation post-transcriptional mRNA alternative splicing and non-coding RNAs and post-translational levels phosphorylation and ubiquitination which may provide novel perspectives for further translational diagnosis or therapeutic strategies targeting hTERT.\\n\\nLabel: True\""
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            }
          },
          "metadata": {},
          "execution_count": 17
        }
      ],
      "source": [
        "train_df[\"text\"][0]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 18,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 156
        },
        "id": "VAHGR_nd4CtH",
        "outputId": "90da7654-a60a-4e56-d432-1d3925c38bb6"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "\"You are a medical text classifier. Your task is to determine if the following text is related to cancer or not.          \\nLabel text only with 'True' or 'False'.          \\nIf text is related to cancer, Label as 'True' otherwise Label as 'False'.          \\nDo not provide any additional explanation.          \\n\\nText:\\nSplenic cyst and its management in a 21-month-old boy: a rare complication of invasive meningococcal disease. Splenic complications of invasive meningococcal disease IMD are well recognised though cyst formation is rare particularly in paediatric populations. The best approach to their management is not yet established. This case outlines the management of a splenic cyst in a 21-month-old boy following severe IMD. The case took place in the context of an acute emergence of serogroup W prompting significant media attention and subsequent change in vaccination practice at a jurisdictional level in Australia. The patient was critically unwell early in the illness then later a collection in the left upper quadrant was detected shown on ultrasound to be a 11.6x7.7 cm splenic cyst. In this case the cyst was managed by ultrasound-guided drainage tube insertion. The residual collection was small and stable on subsequent imaging.\\n\\nLabel: \""
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            }
          },
          "metadata": {},
          "execution_count": 18
        }
      ],
      "source": [
        "test_df[\"text\"][0]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 19,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 156
        },
        "id": "GjGXMMO_5oXG",
        "outputId": "e5ed37ae-2951-4055-e618-408284ec181a"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "\"You are a medical text classifier. Your task is to determine if the following text is related to cancer or not.          \\nLabel text only with 'True' or 'False'.          \\nIf text is related to cancer, Label as 'True' otherwise Label as 'False'.          \\nDo not provide any additional explanation.          \\n\\nText:\\nSimilarities and differences of interstitial lung disease associated with pathogenic variants in SFTPC and ABCA3 in adults. Variants in surfactant genes SFTPC or ABCA3 are responsible for interstitial lung disease ILD in children and adults with few studies in adults. We conducted a multicentre retrospective study of all consecutive adult patients diagnosed with ILD associated with variants in SFTPC or ABCA3 in the French rare pulmonary diseases network OrphaLung. Variants and chest computed tomography CT features were centrally reviewed. We included 36 patients median age: 34 years 20 males 22 in the SFTPC group and 14 in the ABCA3 group. Clinical characteristics were similar between groups. Baseline median FVC was 59 [52-72] and DLco was 44 [35-50]. An unclassifiable pattern of fibrosing ILD was the most frequent on chest CT found in 85 of patients however with a distinct phenotype with ground-glass opacities and/or cysts. Nonspecific interstitial pneumonia and usual interstitial pneumonia were the most common histological patterns in the ABCA3 group and in the SFTPC group respectively. Annually FVC and DL<sub>CO</sub> declined by 1.87 and 2.43 in the SFTPC group respectively and by 0.72 and 0.95 in the ABCA3 group respectively FVC p = 0.014 and DL<sub>CO</sub> p = 0.004 for comparison between groups. Median time to death or lung transplantation was 10 years in the SFTPC group and was not reached at the end of follow-up in the ABCA3 group. SFTPC and ABCA3-associated ILD present with a distinct phenotype and prognosis. A radiologic pattern of fibrosing ILD with ground-glass opacities and/or cysts is frequently found in these rare conditions.\\n\\nLabel: \""
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            }
          },
          "metadata": {},
          "execution_count": 19
        }
      ],
      "source": [
        "val_df[\"text\"][0]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 20,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "3xpEGkzaY9DH",
        "outputId": "8621921b-e632-4fb3-e370-13082186cc63"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Maximum number of tokens in text: 890\n"
          ]
        }
      ],
      "source": [
        "# Using list comprehension to count words in text\n",
        "word_counts_text= [len(Title.split()) for Title in train_df[\"text\"]]\n",
        "# Get the maximum number of words\n",
        "word_counts_text = max(word_counts_text)\n",
        "print(f\"Maximum number of tokens in text: {word_counts_text}\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "WJ6an5iGs_uT"
      },
      "source": [
        "convert into dictionary (expected input for model training )"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 21,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "5lmNlQ3QtF9v",
        "outputId": "3f2f7324-ea29-4551-e4cd-6e9a7f861bb3"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "{'ID': '31186051', 'labels': True, 'text': \"You are a medical text classifier. Your task is to determine if the following text is related to cancer or not.          \\nLabel text only with 'True' or 'False'.          \\nIf text is related to cancer, Label as 'True' otherwise Label as 'False'.          \\nDo not provide any additional explanation.          \\n\\nText:\\nDiverse regulatory manners of human telomerase reverse transcriptase. Human telomerase reverse transcriptase hTERT is the core subunit of human telomerase and plays important roles in human cancers. Aberrant expression of hTERT is closely associated with tumorigenesis cancer cell stemness maintaining cell proliferation apoptosis inhibition senescence evasion and metastasis. The molecular basis of hTERT regulation is highly complicated and consists of various layers. A deep and full-scale comprehension of the regulatory mechanisms of hTERT is pivotal in understanding the pathogenesis and searching for therapeutic approaches. In this review we summarize the recent advances regarding the diverse regulatory mechanisms of hTERT including the transcriptional promoter mutation promoter region methylation and histone acetylation post-transcriptional mRNA alternative splicing and non-coding RNAs and post-translational levels phosphorylation and ubiquitination which may provide novel perspectives for further translational diagnosis or therapeutic strategies targeting hTERT.\\n\\nLabel: True\"}\n"
          ]
        }
      ],
      "source": [
        "##convert into dictionary\n",
        "train_df_dict = train_df.to_dict(orient='list') #Convert df to dictionary\n",
        "test_df_dict = test_df.to_dict(orient='list') #Convert df to dictionary\n",
        "val_df_dict = val_df.to_dict(orient='list') #Convert df to dictionary\n",
        "\n",
        "\n",
        "\n",
        "from datasets import Dataset\n",
        "import pandas as pd\n",
        "\n",
        "# Dataset with values as dictionary of col , val\n",
        "training_dataset = Dataset.from_pandas( pd.DataFrame(train_df_dict))\n",
        "test_dataset = Dataset.from_pandas( pd.DataFrame(test_df_dict))\n",
        "val_dataset = Dataset.from_pandas( pd.DataFrame(val_df_dict))\n",
        "\n",
        "print(training_dataset[0])"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "K1HB69JbydY8"
      },
      "source": [
        "# Model evaluation before fine-tuning"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 22,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "collapsed": true,
        "id": "WMCwY5iztK9G",
        "outputId": "d77f6e47-7aff-45af-d541-61b83043c862"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Processing: 100%|██████████| 200/200 [02:02<00:00,  1.63it/s]\n"
          ]
        }
      ],
      "source": [
        "from transformers import pipeline, logging\n",
        "\n",
        "\n",
        "def predict(test, model, tokenizer):\n",
        "    y_pred = []\n",
        "    categories = [\"True\", \"False\"]\n",
        "\n",
        "    # Assuming 'test' is a list or iterable that you're looping through\n",
        "    total_iterations = len(test)\n",
        "    update_interval = max(1, total_iterations // 5)  # Update every 20% (minimum 1)\n",
        "\n",
        "    # Suppress Transformers warnings and info messages\n",
        "    logging.set_verbosity_error()  # Suppress warnings and informational messages\n",
        "\n",
        "    with tqdm(total=total_iterations, desc=\"Processing\") as pbar:\n",
        "      for i in (range(len(test))):\n",
        "      # for i in (range(5)):\n",
        "\n",
        "\n",
        "        prompt = test_df.iloc[i][\"text\"]\n",
        "        # print(\"--test promp--\")\n",
        "        # print(prompt)\n",
        "        # print(\"----\")\n",
        "        pipe = pipeline(task=\"text-generation\",\n",
        "                        model=model,\n",
        "                        tokenizer=tokenizer,\n",
        "                        max_new_tokens=3,\n",
        "                        temperature=0.1)\n",
        "\n",
        "        result = pipe(prompt)\n",
        "        answer = result[0]['generated_text'].split(\"Label:\")[-1].strip()\n",
        "        # answer = result[0]['generated_text'].split(\"Answer:\")[-1].strip()\n",
        "\n",
        "        # print(\"--prediction--\")\n",
        "        # print(result[0]['generated_text'])\n",
        "        # print(\"----\")\n",
        "        # Determine the predicted category\n",
        "        for category in categories:\n",
        "            if category.lower() in answer.lower():\n",
        "                y_pred.append(category)\n",
        "                break\n",
        "        else:\n",
        "            y_pred.append(\"none\")\n",
        "            # print(result[0]['generated_text'])\n",
        "\n",
        "          # Update every 20%\n",
        "        if (i + 1) % update_interval == 0 or (i + 1) == total_iterations:\n",
        "            pbar.update(update_interval)  # Increment progress bar\n",
        "\n",
        "\n",
        "    return y_pred\n",
        "\n",
        "y_pred = predict(test_df, model, tokenizer)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "sJ88Nax7aMTU"
      },
      "outputs": [],
      "source": [
        "# # Convert to a Pandas Series (for easy counting)\n",
        "# y_pred_series = pd.Series(y_pred)\n",
        "\n",
        "# # Count the occurrences of each value\n",
        "# value_counts = y_pred_series.value_counts()\n",
        "\n",
        "# # Print the counts\n",
        "# print(\"Value Counts:\")\n",
        "# print(value_counts)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 23,
      "metadata": {
        "id": "KO63N8PfLoaW"
      },
      "outputs": [],
      "source": [
        "from sklearn.metrics import (accuracy_score,\n",
        "                             classification_report,\n",
        "                             confusion_matrix)\n",
        "from sklearn.model_selection import train_test_split\n",
        "\n",
        "\n",
        "def evaluate(y_true, y_pred):\n",
        "    labels = [\"True\", \"False\"]\n",
        "    mapping = {'True': 1, 'False': 0}\n",
        "    def map_func(x):\n",
        "        return mapping.get(x, 1)\n",
        "\n",
        "    y_true = np.vectorize(map_func)(y_true)\n",
        "    y_pred = np.vectorize(map_func)(y_pred)\n",
        "\n",
        "    # Calculate accuracy\n",
        "    accuracy = accuracy_score(y_true=y_true, y_pred=y_pred)\n",
        "    print(f'Accuracy: {accuracy:.3f}')\n",
        "\n",
        "    # Generate accuracy report\n",
        "    unique_labels = set(y_true)  # Get unique labels\n",
        "\n",
        "    for label in unique_labels:\n",
        "        label_indices = [i for i in range(len(y_true))\n",
        "                         if y_true[i] == label]\n",
        "        label_y_true = [y_true[i] for i in label_indices]\n",
        "        label_y_pred = [y_pred[i] for i in label_indices]\n",
        "        accuracy = accuracy_score(label_y_true, label_y_pred)\n",
        "        print(f'Accuracy for label {label}: {accuracy:.3f}')\n",
        "\n",
        "    # Generate classification report\n",
        "    class_report = classification_report(y_true=y_true, y_pred=y_pred)\n",
        "    print('\\nClassification Report:')\n",
        "    print(class_report)\n",
        "\n",
        "    # Generate confusion matrix\n",
        "    conf_matrix = confusion_matrix(y_true=y_true, y_pred=y_pred, labels=[0, 1, 2])\n",
        "    print('\\nConfusion Matrix:')\n",
        "    print(conf_matrix)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 24,
      "metadata": {
        "id": "ES8lVSLaMYE_"
      },
      "outputs": [],
      "source": [
        "y_true = test_df.loc[:,'labels']"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 25,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "FWDJskM9MgO2",
        "outputId": "6c710814-8145-48ab-bea7-03d2d3716294"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Accuracy: 0.480\n",
            "Accuracy for label 1: 0.480\n",
            "\n",
            "Classification Report:\n",
            "              precision    recall  f1-score   support\n",
            "\n",
            "           0       0.00      0.00      0.00         0\n",
            "           1       1.00      0.48      0.65       200\n",
            "\n",
            "    accuracy                           0.48       200\n",
            "   macro avg       0.50      0.24      0.32       200\n",
            "weighted avg       1.00      0.48      0.65       200\n",
            "\n",
            "\n",
            "Confusion Matrix:\n",
            "[[  0   0   0]\n",
            " [104  96   0]\n",
            " [  0   0   0]]\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.\n",
            "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n",
            "/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.\n",
            "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n",
            "/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.\n",
            "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n"
          ]
        }
      ],
      "source": [
        "logging.set_verbosity_error()  # Suppress warnings and informational messages\n",
        "evaluate(y_true, y_pred)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "t_uaWp7DCLsD"
      },
      "source": [
        "# LORA Hyper-Parameter tuning (optional)\n",
        "## ( r, lora_alpha, lora_dropout, target_modules )"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "N_KYYNhADwYH"
      },
      "outputs": [],
      "source": [
        "!pip install -q optuna"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "7IwWG1gBCKbT"
      },
      "outputs": [],
      "source": [
        "import optuna\n",
        "from accelerate import Accelerator\n",
        "from transformers import AutoModelForCausalLM, AutoTokenizer, Trainer, TrainingArguments\n",
        "from datasets import load_dataset\n",
        "from datasets import DatasetDict\n",
        "import time\n",
        "\n",
        "\n",
        "start_time = time.time()\n",
        "\n",
        "\n",
        "# Define the objective function\n",
        "def objective(trial):\n",
        "    # Clear GPU cache before loading the model for the second time\n",
        "    torch.cuda.empty_cache()\n",
        "\n",
        "    num_epochs = 1  # desired number of epochs\n",
        "    # batch_size = 1  # per_device_train_batch_size\n",
        "\n",
        "\n",
        "    # TRAIN_DATA_RECORD_SIZE = 10  # size of train/val dataset\n",
        "    # VAL_DATA_RECORD_SIZE = 3\n",
        "    # train_dataset = training_dataset.select(range(TRAIN_DATA_RECORD_SIZE))\n",
        "    # validation_dataset = val_dataset.select(range(VAL_DATA_RECORD_SIZE))\n",
        "    train_dataset = training_dataset\n",
        "    validation_dataset = val_dataset\n",
        "\n",
        "\n",
        "\n",
        "    # training_dataset = train_df\n",
        "    # val_dataset = val_df\n",
        "\n",
        "    # Define hyperparameters to tune\n",
        "    lora_combination = trial.suggest_categorical(\"lora_combination\", [ (32, 16), (32,32),(64, 32)])\n",
        "    lora_r, lora_alpha = lora_combination\n",
        "    lora_dropout = trial.suggest_categorical(\n",
        "        \"lora_dropout\", [0.2,0.3, 0.4]\n",
        "    )  # Higher Rates for smaller dataset or when you observe signs of overfitting during training\n",
        "\n",
        "\n",
        "    target_modules = trial.suggest_categorical(\n",
        "        \"target_modules\",\n",
        "        [\n",
        "            [\"q_proj\", \"v_proj\"],\n",
        "            [\"q_proj\", \"k_proj\", \"v_proj\"],\n",
        "            [\n",
        "                \"q_proj\",\n",
        "                \"o_proj\",\n",
        "                \"k_proj\",\n",
        "                \"v_proj\",\n",
        "                \"gate_proj\",\n",
        "                \"up_proj\",\n",
        "                \"down_proj\",\n",
        "            ],\n",
        "        ],\n",
        "    )\n",
        "\n",
        "\n",
        "    lora_config = LoraConfig(\n",
        "        r=lora_r,  # hyperparam tuning\n",
        "        lora_alpha=lora_alpha,  # hyperparam tuning\n",
        "        lora_dropout=lora_dropout,  # hyperparam tuning\n",
        "        # target_modules=target_modules, # hyperparam tuning\n",
        "        task_type=\"CAUSAL_LM\",\n",
        "    )\n",
        "\n",
        "    # Define training arguments\n",
        "    training_arguments = transformers.TrainingArguments(\n",
        "        per_device_train_batch_size=1,\n",
        "        per_device_eval_batch_size=1,\n",
        "        gradient_accumulation_steps=3,  # 4\n",
        "        num_train_epochs=1,\n",
        "        warmup_steps=3,\n",
        "        eval_strategy=\"steps\",\n",
        "        eval_steps=0.2,\n",
        "        # max_steps=15,\n",
        "        learning_rate=1e-4,\n",
        "        weight_decay=1e-2,  # Add weight decay\n",
        "        fp16=True,\n",
        "        bf16=False,\n",
        "        logging_steps=1,\n",
        "        output_dir=\"outputs\",\n",
        "        optim=\"paged_adamw_8bit\",\n",
        "        report_to=None,  # wandb,tensorboard\n",
        "        label_names=[\"labels\"], ## taget column name\n",
        "\n",
        "    )\n",
        "\n",
        "    # Initialize the Accelerator for distributed processing\n",
        "    accelerator = Accelerator()\n",
        "\n",
        "    # Load model\n",
        "    bnb_config = BitsAndBytesConfig(\n",
        "        load_in_4bit=True,\n",
        "        bnb_4bit_quant_type=\"nf4\", # Quantization type (fp4 or nf4)\n",
        "        bnb_4bit_compute_dtype=\"float16\",  # Compute dtype for 4-bit base models\n",
        "        bnb_4bit_use_double_quant=False,\n",
        "        # Enable CPU offloading for specific layers\n",
        "        llm_int8_enable_fp32_cpu_offload=False, # Activate nested quantization for 4-bit base models (double quantization)\n",
        "        )\n",
        "\n",
        "    model = AutoModelForCausalLM.from_pretrained(\n",
        "        model_id,\n",
        "        quantization_config=bnb_config,\n",
        "        device_map=\"auto\",  # Let Transformers automatically decide device placement\n",
        "        use_auth_token=os.environ[\"HF_TOKEN\"]\n",
        "    )\n",
        "\n",
        "    # Prepare the model, optimizer, and datasets with the Accelerator\n",
        "    model, train_dataset, validation_dataset = accelerator.prepare(\n",
        "        model, train_dataset, validation_dataset\n",
        "    )\n",
        "\n",
        "\n",
        "    # Initialize the Trainer\n",
        "    tokenizer.pad_token = tokenizer.eos_token  # Ensure pad token is set\n",
        "    tokenizer.padding_side = \"left\"  # it is a decoder-only model, it is generally recommended to set padding_side to \"left\".\n",
        "    trainer = SFTTrainer(\n",
        "        model=model,\n",
        "        processing_class=tokenizer,\n",
        "        # data_collator=transformers.DataCollatorForLanguageModeling(tokenizer, mlm=False ,  return_tensors=\"pt\"),\n",
        "        train_dataset=train_dataset,\n",
        "        eval_dataset=validation_dataset,\n",
        "        # max_seq_length=1000,  ## max seq length to input/output. It is crucial for GPU memory management.\n",
        "        dataset_text_field=\"text\",\n",
        "        args=training_arguments,\n",
        "        peft_config=lora_config,\n",
        "        formatting_func=lambda example: example['text'],  # preprocessing function before input\n",
        "\n",
        "    )\n",
        "\n",
        "\n",
        "\n",
        "    # Train the model\n",
        "    trainer.train()\n",
        "\n",
        "\n",
        "    # Evaluate the model\n",
        "    eval_results = trainer.evaluate()\n",
        "\n",
        "\n",
        "\n",
        "    # Return the evaluation metric to optimize\n",
        "    return eval_results[\"eval_loss\"]\n",
        "\n",
        "\n",
        "# Create an Optuna study and optimize the objective function\n",
        "study = optuna.create_study(direction=\"minimize\")\n",
        "study.optimize(objective, n_trials=10)\n",
        "\n",
        "# Print the best hyperparameters\n",
        "best_params = study.best_params\n",
        "print(\"Best hyperparameters: \", best_params)\n",
        "# Print the best performance metrics\n",
        "best_trial = study.best_trial\n",
        "\n",
        "end_time = time.time()\n",
        "print(\"\\n\\n--->Execution Time:\", (end_time - start_time) / 60, \"minutes\")\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "7CnQlBL1M4nf"
      },
      "source": [
        "# Finetune Model"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 26,
      "metadata": {
        "id": "sd1G2rytM6rK"
      },
      "outputs": [],
      "source": [
        "################################################################################\n",
        "# set best LORA parameters\n",
        "# Modlues:\n",
        "# up_proj: Up projection layer, likely part of the model’s feed-forward network.\n",
        "# q_proj: Query projection layer, used in the attention mechanism.\n",
        "# down_proj: Down projection layer, often used after attention or feed-forward layers.\n",
        "# gate_proj: Gating projection layer, possibly used in gated feed-forward networks.\n",
        "# o_proj: Output projection layer, used in the attention mechanism.\n",
        "# k_proj: Key projection layer, used in the attention mechanism.\n",
        "# v_proj: Value projection layer, used in the attention mechanism.\n",
        "################################################################################\n",
        "################################################################################\n",
        "best_lora_dropout =0.4\n",
        "best_lora_r = 32\n",
        "best_lora_alpha =8\n",
        "best_target_modules = [\n",
        "    \"q_proj\",\n",
        "    \"k_proj\",\n",
        "    \"v_proj\",\n",
        "\n",
        "    \"o_proj\",\n",
        "    \"gate_proj\",\n",
        "    \"up_proj\",\n",
        "    \"down_proj\",\n",
        "\n",
        "]\n",
        "\n",
        "# Define LoRA configuration with the best hyperparameters\n",
        "lora_config = LoraConfig(\n",
        "    r=best_lora_r,\n",
        "    lora_alpha=best_lora_alpha,\n",
        "    lora_dropout=best_lora_dropout,\n",
        "    target_modules=best_target_modules,\n",
        "    task_type=\"CAUSAL_LM\",\n",
        "\n",
        ")\n",
        "\n",
        "\n",
        "# Define training arguments\n",
        "training_arguments = transformers.TrainingArguments(\n",
        "  per_device_train_batch_size=1,\n",
        "  per_device_eval_batch_size=1,\n",
        "  gradient_accumulation_steps=3,  # 4\n",
        "  gradient_checkpointing=True,\n",
        "  num_train_epochs=1,\n",
        "  # max_steps=25, #25\n",
        "  warmup_steps=3,\n",
        "  eval_strategy=\"steps\",\n",
        "  eval_steps=0.2,\n",
        "  learning_rate=1e-4,\n",
        "  weight_decay=1e-2,  # Add weight decay\n",
        "  fp16=True,\n",
        "  bf16=False,\n",
        "  logging_steps=1,\n",
        "  output_dir=\"outputs_model_training\",\n",
        "  optim=\"paged_adamw_8bit\",\n",
        "  report_to=\"wandb\",  # wandb,tensorboard\n",
        "  label_names=[\"labels\"], ## taget column name\n",
        "\n",
        ")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "NfYeJgnTOPKP"
      },
      "source": [
        "Distributed training using Accelerator"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 27,
      "metadata": {
        "id": "8FPHRh2XOSQt"
      },
      "outputs": [],
      "source": [
        "# from transformers import AdamW\n",
        "from accelerate import Accelerator\n",
        "from torch.optim import AdamW\n",
        "\n",
        "\n",
        "# Initialize the Accelerator\n",
        "accelerator = Accelerator()\n",
        "\n",
        "# Initialize the optimizer\n",
        "optimizer = AdamW(model.parameters(), lr=training_arguments.learning_rate)\n",
        "\n",
        "# Prepare the model, tokenizer, datasets, and optimizer with the Accelerator\n",
        "model, optimizer, training_dataset, val_dataset = accelerator.prepare(\n",
        "    model, optimizer, training_dataset, val_dataset\n",
        ")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 28,
      "metadata": {
        "id": "hfDBBBZ5Ocuc",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000,
          "referenced_widgets": [
            "ce95f9fa36f04e2183f7ce244210b100",
            "eef15b7b0e034e9594761428533c0074",
            "f2ae370cfed0444cb7996f9991422a82",
            "7392202b6fcb47d0b50f3ad0f7e3be2a",
            "c5068ce5bdb940fa9eecae1e35b26f98",
            "864aff5023394ad8a150b39a76b07cb3",
            "c028daeea2644e9bb1f080df15b823c7",
            "57ad734ea7d4436ba1715af5d7cd9560",
            "b8a04309e86b48fe9d535c1194178ae8",
            "10136eaa560b4d2890806dad55aced1b",
            "90d1bfbdbdf84946bd728f00917383b9",
            "00e0856b5af04a0a81bdec8e48ab23bb",
            "f8674219d6b14953ae2a157b542508f7",
            "1f4cd155b57640bbaba6e31d417466af",
            "a3f863860f0d4fe7beb2f0b1b117a02d",
            "90d2dcdccef14158bd05242c3ff2ef4d",
            "bb762a6021244a0686ffbc8a6bc8992f",
            "1ed3b366ceaa4c8498fc79691a275665",
            "b333b77e29e042239b48da3bc7ab387e",
            "41207fddfa1247de9b894913fa47c07d",
            "f61c33bb26014397ad63cf2d0f73734a",
            "c3563a3d30bc46f08ed3cefb48a78f7d"
          ]
        },
        "outputId": "12f8be18-5d5c-4e9d-a0e9-e113a34ced1e"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_deprecation.py:100: FutureWarning: Deprecated argument(s) used in '__init__': dataset_text_field. Will not be supported from version '0.13.0'.\n",
            "\n",
            "Deprecated positional argument(s) used in SFTTrainer, please use the SFTConfig to set these arguments instead.\n",
            "  warnings.warn(message, FutureWarning)\n",
            "/usr/local/lib/python3.11/dist-packages/trl/trainer/sft_trainer.py:309: UserWarning: You didn't pass a `max_seq_length` argument to the SFTTrainer, this will default to 1024\n",
            "  warnings.warn(\n",
            "/usr/local/lib/python3.11/dist-packages/trl/trainer/sft_trainer.py:328: UserWarning: You passed a `dataset_text_field` argument to the SFTTrainer, the value you passed will override the one in the `SFTConfig`.\n",
            "  warnings.warn(\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "Map:   0%|          | 0/300 [00:00<?, ? examples/s]"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "ce95f9fa36f04e2183f7ce244210b100"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "Map:   0%|          | 0/100 [00:00<?, ? examples/s]"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "00e0856b5af04a0a81bdec8e48ab23bb"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.11/dist-packages/trl/trainer/sft_trainer.py:403: UserWarning: You passed a processing_class with `padding_side` not equal to `right` to the SFTTrainer. This might lead to some unexpected behaviour due to overflow issues when training a model in half-precision. You might consider adding `processing_class.padding_side = 'right'` to your code.\n",
            "  warnings.warn(\n",
            "\u001b[34m\u001b[1mwandb\u001b[0m: \u001b[33mWARNING\u001b[0m The `run_name` is currently set to the same value as `TrainingArguments.output_dir`. If this was not intended, please specify a different run name by setting the `TrainingArguments.run_name` parameter.\n",
            "/usr/local/lib/python3.11/dist-packages/torch/_dynamo/eval_frame.py:745: UserWarning: torch.utils.checkpoint: the use_reentrant parameter should be passed explicitly. In version 2.5 we will raise an exception if use_reentrant is not passed. use_reentrant=False is recommended, but if you need to preserve the current default behavior, you can pass use_reentrant=True. Refer to docs for more details on the differences between the two variants.\n",
            "  return fn(*args, **kwargs)\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "{'loss': 2.1287, 'grad_norm': 0.25255975127220154, 'learning_rate': 0.0, 'epoch': 0.01}\n",
            "{'loss': 1.6509, 'grad_norm': 0.2172706574201584, 'learning_rate': 3.3333333333333335e-05, 'epoch': 0.02}\n",
            "{'loss': 2.1835, 'grad_norm': 0.28841230273246765, 'learning_rate': 6.666666666666667e-05, 'epoch': 0.03}\n",
            "{'loss': 2.2474, 'grad_norm': 0.2552964985370636, 'learning_rate': 0.0001, 'epoch': 0.04}\n",
            "{'loss': 2.1865, 'grad_norm': 0.29049667716026306, 'learning_rate': 9.896907216494846e-05, 'epoch': 0.05}\n",
            "{'loss': 1.877, 'grad_norm': 0.21858148276805878, 'learning_rate': 9.793814432989691e-05, 'epoch': 0.06}\n",
            "{'loss': 2.1217, 'grad_norm': 0.27944955229759216, 'learning_rate': 9.690721649484537e-05, 'epoch': 0.07}\n",
            "{'loss': 2.2569, 'grad_norm': 0.22998777031898499, 'learning_rate': 9.587628865979382e-05, 'epoch': 0.08}\n",
            "{'loss': 1.7339, 'grad_norm': 0.22770118713378906, 'learning_rate': 9.484536082474227e-05, 'epoch': 0.09}\n",
            "{'loss': 1.8556, 'grad_norm': 0.2435479760169983, 'learning_rate': 9.381443298969073e-05, 'epoch': 0.1}\n",
            "{'loss': 1.8096, 'grad_norm': 0.27305272221565247, 'learning_rate': 9.278350515463918e-05, 'epoch': 0.11}\n",
            "{'loss': 1.8238, 'grad_norm': 0.2820190489292145, 'learning_rate': 9.175257731958763e-05, 'epoch': 0.12}\n",
            "{'loss': 1.7436, 'grad_norm': 0.2836551070213318, 'learning_rate': 9.072164948453609e-05, 'epoch': 0.13}\n",
            "{'loss': 1.9412, 'grad_norm': 0.32276520133018494, 'learning_rate': 8.969072164948454e-05, 'epoch': 0.14}\n",
            "{'loss': 1.6176, 'grad_norm': 0.28054124116897583, 'learning_rate': 8.865979381443299e-05, 'epoch': 0.15}\n",
            "{'loss': 1.8039, 'grad_norm': 0.24707254767417908, 'learning_rate': 8.762886597938145e-05, 'epoch': 0.16}\n",
            "{'loss': 1.587, 'grad_norm': 0.256268709897995, 'learning_rate': 8.65979381443299e-05, 'epoch': 0.17}\n",
            "{'loss': 1.7524, 'grad_norm': 0.24219566583633423, 'learning_rate': 8.556701030927835e-05, 'epoch': 0.18}\n",
            "{'loss': 1.6039, 'grad_norm': 0.29597339034080505, 'learning_rate': 8.453608247422681e-05, 'epoch': 0.19}\n",
            "{'loss': 1.7469, 'grad_norm': 0.26856714487075806, 'learning_rate': 8.350515463917527e-05, 'epoch': 0.2}\n",
            "{'eval_loss': 1.7432289123535156, 'eval_runtime': 63.0793, 'eval_samples_per_second': 1.585, 'eval_steps_per_second': 1.585, 'epoch': 0.2}\n",
            "{'loss': 1.5796, 'grad_norm': 0.22787386178970337, 'learning_rate': 8.247422680412371e-05, 'epoch': 0.21}\n",
            "{'loss': 1.7439, 'grad_norm': 0.21623611450195312, 'learning_rate': 8.144329896907217e-05, 'epoch': 0.22}\n",
            "{'loss': 1.5763, 'grad_norm': 0.21448121964931488, 'learning_rate': 8.041237113402063e-05, 'epoch': 0.23}\n",
            "{'loss': 1.6889, 'grad_norm': 0.24175477027893066, 'learning_rate': 7.938144329896907e-05, 'epoch': 0.24}\n",
            "{'loss': 1.7242, 'grad_norm': 0.2773129343986511, 'learning_rate': 7.835051546391753e-05, 'epoch': 0.25}\n",
            "{'loss': 1.5466, 'grad_norm': 0.24184812605381012, 'learning_rate': 7.731958762886599e-05, 'epoch': 0.26}\n",
            "{'loss': 1.7172, 'grad_norm': 0.33574938774108887, 'learning_rate': 7.628865979381443e-05, 'epoch': 0.27}\n",
            "{'loss': 1.8008, 'grad_norm': 0.26655998826026917, 'learning_rate': 7.525773195876289e-05, 'epoch': 0.28}\n",
            "{'loss': 1.6836, 'grad_norm': 0.26892995834350586, 'learning_rate': 7.422680412371135e-05, 'epoch': 0.29}\n",
            "{'loss': 1.8107, 'grad_norm': 0.2732125520706177, 'learning_rate': 7.319587628865979e-05, 'epoch': 0.3}\n",
            "{'loss': 1.5068, 'grad_norm': 0.25205498933792114, 'learning_rate': 7.216494845360825e-05, 'epoch': 0.31}\n",
            "{'loss': 1.7239, 'grad_norm': 0.2548144459724426, 'learning_rate': 7.113402061855671e-05, 'epoch': 0.32}\n",
            "{'loss': 1.6373, 'grad_norm': 0.2602764368057251, 'learning_rate': 7.010309278350515e-05, 'epoch': 0.33}\n",
            "{'loss': 1.6879, 'grad_norm': 0.2324882596731186, 'learning_rate': 6.907216494845361e-05, 'epoch': 0.34}\n",
            "{'loss': 1.4237, 'grad_norm': 0.26809200644493103, 'learning_rate': 6.804123711340207e-05, 'epoch': 0.35}\n",
            "{'loss': 1.5513, 'grad_norm': 0.24896657466888428, 'learning_rate': 6.701030927835051e-05, 'epoch': 0.36}\n",
            "{'loss': 1.7307, 'grad_norm': 0.2566555440425873, 'learning_rate': 6.597938144329897e-05, 'epoch': 0.37}\n",
            "{'loss': 1.6272, 'grad_norm': 0.2915957570075989, 'learning_rate': 6.494845360824743e-05, 'epoch': 0.38}\n",
            "{'loss': 1.4712, 'grad_norm': 0.22831721603870392, 'learning_rate': 6.391752577319587e-05, 'epoch': 0.39}\n",
            "{'loss': 1.7893, 'grad_norm': 0.3259837329387665, 'learning_rate': 6.288659793814433e-05, 'epoch': 0.4}\n",
            "{'eval_loss': 1.6760777235031128, 'eval_runtime': 62.9033, 'eval_samples_per_second': 1.59, 'eval_steps_per_second': 1.59, 'epoch': 0.4}\n",
            "{'loss': 1.6337, 'grad_norm': 0.2779144048690796, 'learning_rate': 6.185567010309279e-05, 'epoch': 0.41}\n",
            "{'loss': 1.423, 'grad_norm': 0.2360135018825531, 'learning_rate': 6.0824742268041234e-05, 'epoch': 0.42}\n",
            "{'loss': 1.6154, 'grad_norm': 0.23614008724689484, 'learning_rate': 5.979381443298969e-05, 'epoch': 0.43}\n",
            "{'loss': 1.8173, 'grad_norm': 0.5735219717025757, 'learning_rate': 5.876288659793815e-05, 'epoch': 0.44}\n",
            "{'loss': 1.4176, 'grad_norm': 0.3447019159793854, 'learning_rate': 5.7731958762886594e-05, 'epoch': 0.45}\n",
            "{'loss': 1.5351, 'grad_norm': 0.22670453786849976, 'learning_rate': 5.670103092783505e-05, 'epoch': 0.46}\n",
            "{'loss': 1.7331, 'grad_norm': 0.2905718982219696, 'learning_rate': 5.567010309278351e-05, 'epoch': 0.47}\n",
            "{'loss': 1.6507, 'grad_norm': 0.3425893783569336, 'learning_rate': 5.4639175257731955e-05, 'epoch': 0.48}\n",
            "{'loss': 1.4759, 'grad_norm': 0.3047894239425659, 'learning_rate': 5.360824742268041e-05, 'epoch': 0.49}\n",
            "{'loss': 1.719, 'grad_norm': 0.27246493101119995, 'learning_rate': 5.257731958762887e-05, 'epoch': 0.5}\n",
            "{'loss': 1.5889, 'grad_norm': 0.2832053303718567, 'learning_rate': 5.1546391752577315e-05, 'epoch': 0.51}\n",
            "{'loss': 1.3733, 'grad_norm': 0.2512577772140503, 'learning_rate': 5.051546391752577e-05, 'epoch': 0.52}\n",
            "{'loss': 1.6913, 'grad_norm': 0.2307763546705246, 'learning_rate': 4.948453608247423e-05, 'epoch': 0.53}\n",
            "{'loss': 1.6412, 'grad_norm': 0.3065043091773987, 'learning_rate': 4.845360824742268e-05, 'epoch': 0.54}\n",
            "{'loss': 1.1129, 'grad_norm': 0.39159855246543884, 'learning_rate': 4.7422680412371134e-05, 'epoch': 0.55}\n",
            "{'loss': 1.757, 'grad_norm': 0.2650320827960968, 'learning_rate': 4.639175257731959e-05, 'epoch': 0.56}\n",
            "{'loss': 1.4952, 'grad_norm': 0.2391061931848526, 'learning_rate': 4.536082474226804e-05, 'epoch': 0.57}\n",
            "{'loss': 1.3672, 'grad_norm': 0.23963436484336853, 'learning_rate': 4.4329896907216494e-05, 'epoch': 0.58}\n",
            "{'loss': 1.8829, 'grad_norm': 0.290272980928421, 'learning_rate': 4.329896907216495e-05, 'epoch': 0.59}\n",
            "{'loss': 1.5275, 'grad_norm': 0.251201868057251, 'learning_rate': 4.2268041237113404e-05, 'epoch': 0.6}\n",
            "{'eval_loss': 1.6578277349472046, 'eval_runtime': 63.3917, 'eval_samples_per_second': 1.577, 'eval_steps_per_second': 1.577, 'epoch': 0.6}\n",
            "{'loss': 1.8344, 'grad_norm': 0.2873877286911011, 'learning_rate': 4.1237113402061855e-05, 'epoch': 0.61}\n",
            "{'loss': 1.5252, 'grad_norm': 0.2542914152145386, 'learning_rate': 4.020618556701031e-05, 'epoch': 0.62}\n",
            "{'loss': 1.5491, 'grad_norm': 0.2655899226665497, 'learning_rate': 3.9175257731958764e-05, 'epoch': 0.63}\n",
            "{'loss': 1.7157, 'grad_norm': 0.2621350586414337, 'learning_rate': 3.8144329896907216e-05, 'epoch': 0.64}\n",
            "{'loss': 1.47, 'grad_norm': 0.26656416058540344, 'learning_rate': 3.7113402061855674e-05, 'epoch': 0.65}\n",
            "{'loss': 1.4261, 'grad_norm': 0.24769830703735352, 'learning_rate': 3.6082474226804125e-05, 'epoch': 0.66}\n",
            "{'loss': 1.4803, 'grad_norm': 0.27293211221694946, 'learning_rate': 3.5051546391752576e-05, 'epoch': 0.67}\n",
            "{'loss': 1.7241, 'grad_norm': 0.2616881728172302, 'learning_rate': 3.4020618556701034e-05, 'epoch': 0.68}\n",
            "{'loss': 1.599, 'grad_norm': 0.3390539586544037, 'learning_rate': 3.2989690721649485e-05, 'epoch': 0.69}\n",
            "{'loss': 1.6568, 'grad_norm': 0.23852714896202087, 'learning_rate': 3.1958762886597937e-05, 'epoch': 0.7}\n",
            "{'loss': 1.6948, 'grad_norm': 0.27562472224235535, 'learning_rate': 3.0927835051546395e-05, 'epoch': 0.71}\n",
            "{'loss': 1.6248, 'grad_norm': 0.3281119465827942, 'learning_rate': 2.9896907216494846e-05, 'epoch': 0.72}\n",
            "{'loss': 1.4673, 'grad_norm': 0.23641806840896606, 'learning_rate': 2.8865979381443297e-05, 'epoch': 0.73}\n",
            "{'loss': 1.7206, 'grad_norm': 0.2653757631778717, 'learning_rate': 2.7835051546391755e-05, 'epoch': 0.74}\n",
            "{'loss': 1.4523, 'grad_norm': 0.2836109697818756, 'learning_rate': 2.6804123711340206e-05, 'epoch': 0.75}\n",
            "{'loss': 1.5613, 'grad_norm': 0.23214749991893768, 'learning_rate': 2.5773195876288658e-05, 'epoch': 0.76}\n",
            "{'loss': 1.3846, 'grad_norm': 0.24832971394062042, 'learning_rate': 2.4742268041237116e-05, 'epoch': 0.77}\n",
            "{'loss': 1.7577, 'grad_norm': 0.30103611946105957, 'learning_rate': 2.3711340206185567e-05, 'epoch': 0.78}\n",
            "{'loss': 1.6758, 'grad_norm': 0.2023988664150238, 'learning_rate': 2.268041237113402e-05, 'epoch': 0.79}\n",
            "{'loss': 1.6729, 'grad_norm': 0.2565455436706543, 'learning_rate': 2.1649484536082476e-05, 'epoch': 0.8}\n",
            "{'eval_loss': 1.6505303382873535, 'eval_runtime': 63.0471, 'eval_samples_per_second': 1.586, 'eval_steps_per_second': 1.586, 'epoch': 0.8}\n",
            "{'loss': 1.4675, 'grad_norm': 0.226011261343956, 'learning_rate': 2.0618556701030927e-05, 'epoch': 0.81}\n",
            "{'loss': 1.4675, 'grad_norm': 0.24572354555130005, 'learning_rate': 1.9587628865979382e-05, 'epoch': 0.82}\n",
            "{'loss': 1.7415, 'grad_norm': 0.262448251247406, 'learning_rate': 1.8556701030927837e-05, 'epoch': 0.83}\n",
            "{'loss': 1.2953, 'grad_norm': 0.2856847047805786, 'learning_rate': 1.7525773195876288e-05, 'epoch': 0.84}\n",
            "{'loss': 1.5353, 'grad_norm': 0.34350818395614624, 'learning_rate': 1.6494845360824743e-05, 'epoch': 0.85}\n",
            "{'loss': 1.8405, 'grad_norm': 0.2508811056613922, 'learning_rate': 1.5463917525773197e-05, 'epoch': 0.86}\n",
            "{'loss': 1.688, 'grad_norm': 0.26456597447395325, 'learning_rate': 1.4432989690721649e-05, 'epoch': 0.87}\n",
            "{'loss': 1.6202, 'grad_norm': 0.34648725390434265, 'learning_rate': 1.3402061855670103e-05, 'epoch': 0.88}\n",
            "{'loss': 1.6203, 'grad_norm': 0.25537940859794617, 'learning_rate': 1.2371134020618558e-05, 'epoch': 0.89}\n",
            "{'loss': 1.6521, 'grad_norm': 0.2476256638765335, 'learning_rate': 1.134020618556701e-05, 'epoch': 0.9}\n",
            "{'loss': 1.5036, 'grad_norm': 0.2601606845855713, 'learning_rate': 1.0309278350515464e-05, 'epoch': 0.91}\n",
            "{'loss': 1.4214, 'grad_norm': 0.2506980001926422, 'learning_rate': 9.278350515463918e-06, 'epoch': 0.92}\n",
            "{'loss': 1.5588, 'grad_norm': 0.23155543208122253, 'learning_rate': 8.247422680412371e-06, 'epoch': 0.93}\n",
            "{'loss': 1.6121, 'grad_norm': 0.26005303859710693, 'learning_rate': 7.216494845360824e-06, 'epoch': 0.94}\n",
            "{'loss': 1.7717, 'grad_norm': 0.28349030017852783, 'learning_rate': 6.185567010309279e-06, 'epoch': 0.95}\n",
            "{'loss': 1.6328, 'grad_norm': 0.28907373547554016, 'learning_rate': 5.154639175257732e-06, 'epoch': 0.96}\n",
            "{'loss': 1.4253, 'grad_norm': 0.2732320725917816, 'learning_rate': 4.123711340206186e-06, 'epoch': 0.97}\n",
            "{'loss': 1.8403, 'grad_norm': 0.27049311995506287, 'learning_rate': 3.0927835051546395e-06, 'epoch': 0.98}\n",
            "{'loss': 1.5261, 'grad_norm': 0.24899867177009583, 'learning_rate': 2.061855670103093e-06, 'epoch': 0.99}\n",
            "{'loss': 1.5411, 'grad_norm': 0.27002233266830444, 'learning_rate': 1.0309278350515464e-06, 'epoch': 1.0}\n",
            "{'eval_loss': 1.6483862400054932, 'eval_runtime': 63.2749, 'eval_samples_per_second': 1.58, 'eval_steps_per_second': 1.58, 'epoch': 1.0}\n",
            "{'train_runtime': 910.4382, 'train_samples_per_second': 0.33, 'train_steps_per_second': 0.11, 'train_loss': 1.6590507590770722, 'epoch': 1.0}\n",
            "saving model Llama-3.1-8B-Inst_cancer_classification_finetuned\n",
            "\n",
            "\n",
            "--->Execution Time: 15.257585481802623 minutes\n"
          ]
        }
      ],
      "source": [
        "from accelerate import DistributedType\n",
        "import time\n",
        "\n",
        "\n",
        "\n",
        "start_time = time.time()\n",
        "\n",
        "# Clear GPU cache before loading the model for the second time\n",
        "torch.cuda.empty_cache()\n",
        "\n",
        "\n",
        "SAVE_MODEL = True\n",
        "\n",
        "tokenizer.pad_token = tokenizer.eos_token  # Ensure pad token is set\n",
        "#If you pad on the right and haven't shifted labels appropriately, the model might try to predict padding tokens as the \"next\" word, which is incorrect\n",
        "tokenizer.padding_side = \"left\"  # it is a decoder-only model, it is generally recommended to set padding_side to \"left\".\n",
        "\n",
        "\n",
        "trainer = SFTTrainer(\n",
        "    model=model,\n",
        "    processing_class=tokenizer,\n",
        "    train_dataset=training_dataset.select(range(300)), #<<--- limiting due to time,compute constraint\n",
        "    eval_dataset=val_dataset.select(range(100)),\n",
        "    # max_seq_length=1000,  ## max seq length to input/output. It is crucial for GPU memory management.\n",
        "    dataset_text_field=\"text\",\n",
        "    args=training_arguments,\n",
        "    peft_config=lora_config,\n",
        "    formatting_func=lambda example: example['text'],  # preprocessing function before input\n",
        "    data_collator=transformers.DataCollatorForLanguageModeling(tokenizer, mlm=False ,  return_tensors=\"pt\"),\n",
        "\n",
        ")\n",
        "\n",
        "model.config.use_cache = False\n",
        "\n",
        "# Train the final model\n",
        "trainer.train()\n",
        "\n",
        "\n",
        "# Save the final model\n",
        "# accelerator.wait_for_everyone() method is used to synchronize all processes in a distributed training setup,ensuring that all processes reach the same point before proceeding.\n",
        "# This is crucial for maintaining consistency and coordination across multiple devices (e.g., multiple GPUs or TPUs) during training.\n",
        "accelerator.wait_for_everyone() # Use the Accelerator to manage the training loop\n",
        "if accelerator.is_local_main_process:\n",
        "    if SAVE_MODEL:\n",
        "        print(f\"saving model {new_model}\")\n",
        "        trainer.model.save_pretrained(new_model)\n",
        "        trainer.tokenizer.save_pretrained(new_model)\n",
        "\n",
        "end_time = time.time()\n",
        "print(\"\\n\\n--->Execution Time:\", (end_time - start_time) / 60, \"minutes\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "u7Erm2-NS37H"
      },
      "source": [
        "# Evaluate model after finetuning"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 29,
      "metadata": {
        "id": "HN8ZZZLJSCem",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "0d57f6bc-0f8e-45fa-f642-66fae8b7a540"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Processing: 100%|██████████| 200/200 [02:49<00:00,  1.18it/s]\n"
          ]
        }
      ],
      "source": [
        "y_pred = predict(test_df, model, tokenizer)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "6MUyVPMJHKSa"
      },
      "outputs": [],
      "source": [
        "# # Convert to a Pandas Series (for easy counting)\n",
        "# y_pred_series = pd.Series(y_pred)\n",
        "\n",
        "# # Count the occurrences of each value\n",
        "# value_counts = y_pred_series.value_counts()\n",
        "\n",
        "# # Print the counts\n",
        "# print(\"Value Counts:\")\n",
        "# print(value_counts)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 30,
      "metadata": {
        "id": "FsR9cSnrX9zn",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "ee7c9544-3f89-41c2-8a29-211143b5f5df"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Accuracy: 0.760\n",
            "Accuracy for label 1: 0.760\n",
            "\n",
            "Classification Report:\n",
            "              precision    recall  f1-score   support\n",
            "\n",
            "           0       0.00      0.00      0.00         0\n",
            "           1       1.00      0.76      0.86       200\n",
            "\n",
            "    accuracy                           0.76       200\n",
            "   macro avg       0.50      0.38      0.43       200\n",
            "weighted avg       1.00      0.76      0.86       200\n",
            "\n",
            "\n",
            "Confusion Matrix:\n",
            "[[  0   0   0]\n",
            " [ 48 152   0]\n",
            " [  0   0   0]]\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.\n",
            "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n",
            "/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.\n",
            "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n",
            "/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.\n",
            "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n"
          ]
        }
      ],
      "source": [
        "evaluate(y_true, y_pred)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 31,
      "metadata": {
        "id": "lZiBODIFPRPM",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 765
        },
        "outputId": "a68c2f88-cfcc-4a27-d219-52f95a48398f"
      },
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<br>    <style><br>        .wandb-row {<br>            display: flex;<br>            flex-direction: row;<br>            flex-wrap: wrap;<br>            justify-content: flex-start;<br>            width: 100%;<br>        }<br>        .wandb-col {<br>            display: flex;<br>            flex-direction: column;<br>            flex-basis: 100%;<br>            flex: 1;<br>            padding: 10px;<br>        }<br>    </style><br><div class=\"wandb-row\"><div class=\"wandb-col\"><h3>Run history:</h3><br/><table class=\"wandb\"><tr><td>eval/loss</td><td>█▃▂▁▁</td></tr><tr><td>eval/runtime</td><td>▄▁█▃▆</td></tr><tr><td>eval/samples_per_second</td><td>▅█▁▆▃</td></tr><tr><td>eval/steps_per_second</td><td>▅█▁▆▃</td></tr><tr><td>train/epoch</td><td>▁▁▁▂▂▂▂▃▃▃▃▃▃▄▄▄▅▅▅▅▆▆▆▆▆▆▆▆▇▇▇▇▇▇▇▇▇███</td></tr><tr><td>train/global_step</td><td>▁▁▂▂▂▂▂▂▂▂▃▃▃▃▃▄▄▄▄▄▄▄▅▅▅▅▅▆▆▆▇▇▇▇▇█████</td></tr><tr><td>train/grad_norm</td><td>▁▅▂▃▄▃▂▅▄▁█▄▄▃▃▂▃▂▇▄█▂▆▅▃▂▅▃▄▄▄█▂▃▂▅█▃▃▄</td></tr><tr><td>train/learning_rate</td><td>████▇▇▇▇▇▇▆▆▆▆▆▆▆▅▅▅▅▅▅▄▄▄▄▄▃▃▃▃▃▂▂▂▂▁▁▁</td></tr><tr><td>train/loss</td><td>█▅█▆█▆▅▄▅▄▄▅▄▅▄▄▄▅▅▅▄▁▃▃▆▄▅▅▅▄▄▃▅▅▃▂▄▆▅▄</td></tr></table><br/></div><div class=\"wandb-col\"><h3>Run summary:</h3><br/><table class=\"wandb\"><tr><td>eval/loss</td><td>1.64839</td></tr><tr><td>eval/runtime</td><td>63.2749</td></tr><tr><td>eval/samples_per_second</td><td>1.58</td></tr><tr><td>eval/steps_per_second</td><td>1.58</td></tr><tr><td>total_flos</td><td>5525367575666688.0</td></tr><tr><td>train/epoch</td><td>1</td></tr><tr><td>train/global_step</td><td>100</td></tr><tr><td>train/grad_norm</td><td>0.27002</td></tr><tr><td>train/learning_rate</td><td>0.0</td></tr><tr><td>train/loss</td><td>1.5411</td></tr><tr><td>train_loss</td><td>1.65905</td></tr><tr><td>train_runtime</td><td>910.4382</td></tr><tr><td>train_samples_per_second</td><td>0.33</td></tr><tr><td>train_steps_per_second</td><td>0.11</td></tr></table><br/></div></div>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              " View run <strong style=\"color:#cdcd00\">pretty-star-33</strong> at: <a href='https://wandb.ai/pratik_ai/cancer_classification/runs/ko349hw3?apiKey=9d7decf681236b200a35c0121bca0fe725be724c' target=\"_blank\">https://wandb.ai/pratik_ai/cancer_classification/runs/ko349hw3?apiKey=9d7decf681236b200a35c0121bca0fe725be724c</a><br> View project at: <a href='https://wandb.ai/pratik_ai/cancer_classification?apiKey=9d7decf681236b200a35c0121bca0fe725be724c' target=\"_blank\">https://wandb.ai/pratik_ai/cancer_classification?apiKey=9d7decf681236b200a35c0121bca0fe725be724c</a><br>Synced 5 W&B file(s), 0 media file(s), 0 artifact file(s) and 0 other file(s)"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "Find logs at: <code>./wandb/run-20250601_121041-ko349hw3/logs</code>"
            ]
          },
          "metadata": {}
        }
      ],
      "source": [
        "wandb.finish()\n",
        "model.config.use_cache = True"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "t3Iwvtk6UDvC"
      },
      "source": [
        "# Merge finetuned LORA with pre-trained model\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Q5kKRsdLUCii"
      },
      "outputs": [],
      "source": [
        "# Clear GPU cache before loading the model for the second time\n",
        "torch.cuda.empty_cache()\n",
        "\n",
        "\n",
        "from peft import LoraConfig, PeftModel\n",
        "\n",
        "base_model = AutoModelForCausalLM.from_pretrained(\n",
        "    model_id,\n",
        "    quantization_config=bnb_config,\n",
        "    low_cpu_mem_usage=True,\n",
        "    return_dict=True,\n",
        "    torch_dtype=torch.float16,\n",
        "    device_map=\"auto\",\n",
        ")\n",
        "model = PeftModel.from_pretrained(base_model, new_model)\n",
        "finetuned_model = model.merge_and_unload()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "jANUFe6WSEsL"
      },
      "source": [
        "# Push Model to Huggingface hub\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "LDa-EVWkSFvo"
      },
      "outputs": [],
      "source": [
        "trainer.model.push_to_hub(new_model, use_temp_dir=False)"
      ]
    }
  ],
  "metadata": {
    "accelerator": "GPU",
    "colab": {
      "gpuType": "T4",
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }

  },
  "nbformat": 4,
  "nbformat_minor": 0
}
